<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.549">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Working with Snakemake</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../assets/favicon.png" rel="icon" type="image/png">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<style>html{ scroll-behavior: smooth; }</style>
<style>

      .quarto-title-block .quarto-title-banner {
        background-image: url(../assets/images/banner.webp);
background-size: cover;
      }
</style>
<link rel="preconnect" href="https://fonts.googleapis.com">
<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin="">
<link href="https://fonts.googleapis.com/css2?family=Fira+Mono&amp;family=Nunito:ital,wght@0,400;0,500;0,600;1,400;1,500;1,600&amp;display=swap" rel="stylesheet">


</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top quarto-banner">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../assets/logos/nbis-scilifelab.png" alt="logo" class="navbar-logo">
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../home_schedule.html"> 
<span class="menu-text">Schedule</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../home_contents.html"> 
<span class="menu-text">Contents</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../home_syllabus.html"> 
<span class="menu-text">Syllabus</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../home_precourse.html"> 
<span class="menu-text">Pre-course</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../home_info.html"> 
<span class="menu-text">Info</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/NBISweden/workshop-reproducible-research/"> <i class="bi bi-github" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">Working with Snakemake</h1>
            <p class="subtitle lead">How to create reproducible workflows and computational pipelines</p>
                      </div>
  </div>
    
  
  <div class="quarto-title-meta">

      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">18-Oct-2024</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction"><span class="header-section-number">1</span> Introduction</a></li>
  <li><a href="#the-basics" id="toc-the-basics" class="nav-link" data-scroll-target="#the-basics"><span class="header-section-number">2</span> The basics</a></li>
  <li><a href="#visualising-workflows" id="toc-visualising-workflows" class="nav-link" data-scroll-target="#visualising-workflows"><span class="header-section-number">3</span> Visualising workflows</a></li>
  <li><a href="#the-mrsa-workflow" id="toc-the-mrsa-workflow" class="nav-link" data-scroll-target="#the-mrsa-workflow"><span class="header-section-number">4</span> The MRSA workflow</a></li>
  <li><a href="#parameters" id="toc-parameters" class="nav-link" data-scroll-target="#parameters"><span class="header-section-number">5</span> Parameters</a></li>
  <li><a href="#logs" id="toc-logs" class="nav-link" data-scroll-target="#logs"><span class="header-section-number">6</span> Logs</a></li>
  <li><a href="#temporary-files" id="toc-temporary-files" class="nav-link" data-scroll-target="#temporary-files"><span class="header-section-number">7</span> Temporary files</a></li>
  <li><a href="#targets" id="toc-targets" class="nav-link" data-scroll-target="#targets"><span class="header-section-number">8</span> Targets</a></li>
  <li><a href="#shadow-rules" id="toc-shadow-rules" class="nav-link" data-scroll-target="#shadow-rules"><span class="header-section-number">9</span> Shadow rules</a></li>
  <li><a href="#generalising-workflows" id="toc-generalising-workflows" class="nav-link" data-scroll-target="#generalising-workflows"><span class="header-section-number">10</span> Generalising workflows</a></li>
  <li><a href="#reading-samples-from-a-file-instead-of-hard-coding-them" id="toc-reading-samples-from-a-file-instead-of-hard-coding-them" class="nav-link" data-scroll-target="#reading-samples-from-a-file-instead-of-hard-coding-them"><span class="header-section-number">11</span> Reading samples from a file instead of hard-coding them</a></li>
  <li><a href="#extra-material" id="toc-extra-material" class="nav-link" data-scroll-target="#extra-material"><span class="header-section-number">12</span> Extra material</a>
  <ul>
  <li><a href="#using-containers-in-snakemake" id="toc-using-containers-in-snakemake" class="nav-link" data-scroll-target="#using-containers-in-snakemake"><span class="header-section-number">12.1</span> Using containers in Snakemake</a></li>
  <li><a href="#running-snakemake-workflows-on-hpc-clusters" id="toc-running-snakemake-workflows-on-hpc-clusters" class="nav-link" data-scroll-target="#running-snakemake-workflows-on-hpc-clusters"><span class="header-section-number">12.2</span> Running Snakemake workflows on HPC clusters</a>
  <ul class="collapse">
  <li><a href="#option-1-run-the-entire-workflow-as-a-single-job" id="toc-option-1-run-the-entire-workflow-as-a-single-job" class="nav-link" data-scroll-target="#option-1-run-the-entire-workflow-as-a-single-job"><span class="header-section-number">12.2.1</span> Option 1: Run the entire workflow as a single job</a></li>
  <li><a href="#option-2-use-built-in-slurm-support" id="toc-option-2-use-built-in-slurm-support" class="nav-link" data-scroll-target="#option-2-use-built-in-slurm-support"><span class="header-section-number">12.2.2</span> Option 2: Use built-in SLURM support</a></li>
  </ul></li>
  <li><a href="#specifying-resources-for-slurm" id="toc-specifying-resources-for-slurm" class="nav-link" data-scroll-target="#specifying-resources-for-slurm"><span class="header-section-number">12.3</span> Specifying resources for SLURM</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">






<section id="introduction" class="level2" data-number="1">
<h2 data-number="1" class="anchored" data-anchor-id="introduction"><span class="header-section-number">1</span> Introduction</h2>
<p>A <em>workflow management system</em> (WfMS) is a piece of software that sets up, performs and monitors a defined sequence of computational tasks (<em>i.e.</em> “a workflow”). Snakemake is a WfMS that was developed in the bioinformatics community, and as such it has a number of features that make it particularly well-suited for creating reproducible and scalable data analyses.</p>
<p>First of all the language you use to formulate your workflows is based on Python, which is a language with strong standing in academia. However, users are not required to know how to code in Python to work efficiently with Snakemake. Workflows can easily be scaled from your desktop to server, cluster, grid or cloud environments. This makes it possible to develop a workflow on your laptop, maybe using only a small subset of your data, and then run the real analysis on a cluster. Snakemake also has several features for defining the environment with which each task is carried out. This is important in bioinformatics, where workflows often involve running a large number of small third-party tools.</p>
<p>Snakemake is primarily intended to work on <em>files</em> (rather than for example streams, reading/writing from databases or passing variables in memory). This fits well with many fields of bioinformatics, notably next-generation sequencing, that often involve computationally expensive operations on large files. It’s also a good fit for a scientific research setting, where the exact specifications of the final workflow aren’t always known at the beginning of a project.</p>
<p>Lastly, a WfMS is a very important tool for making your analyses reproducible. By keeping track of when each file was generated, and by which operation, it is possible to ensure that there is a consistent “paper trail” from raw data to final results. Snakemake also has features that allow you to package and distribute the workflow, and any files it involves, once it’s done.</p>
<p>This tutorial depends on files from the course GitHub repo. Take a look at the <a href="pre-course-setup">setup</a> for instructions on how to set it up if you haven’t done so already, then open up a terminal and go to <code>workshop-reproducible-research/tutorials/snakemake</code> and activate your <code>snakemake-env</code> Conda environment.</p>
</section>
<section id="the-basics" class="level2" data-number="2">
<h2 data-number="2" class="anchored" data-anchor-id="the-basics"><span class="header-section-number">2</span> The basics</h2>
<p>In this part of the tutorial we will create a very simple workflow from scratch, in order to show the fundamentals of how Snakemake works. The workflow will take two files as inputs, <code>a.txt</code> and <code>b.txt</code>, and the purpose is to convert the text in the files to upper case and then to concatenate them.</p>
<p>Run the following shell commands. The first one will make an empty file named <code>Snakefile</code>, which will later contain the workflow. The second and third commands generate two files containing some arbitrary text.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">touch</span> Snakefile</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"This is a.txt"</span> <span class="op">&gt;</span> a.txt</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"This is b.txt"</span> <span class="op">&gt;</span> b.txt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Then open <code>Snakefile</code> in your favourite text editor. A Snakemake workflow is based on rules which take some file(s) as input, performs some type of operation on them, and generate some file(s) as outputs. Here is a very simple rule that produces <code>a.upper.txt</code> as an output, using <code>a.txt</code> as input. Copy this rule to your <code>Snakefile</code> and save it.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>rule convert_to_upper_case:</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>    output:</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>        <span class="co">"a.upper.txt"</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>    <span class="bu">input</span>:</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>        <span class="co">"a.txt"</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>    shell:</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="co">        tr [a-z] [A-Z] &lt; {input} &gt; {output}</span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a><span class="co">        """</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="callout callout-style-default callout-caution callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Caution
</div>
</div>
<div class="callout-body-container callout-body">
<p>Indentation is important in Snakefiles, so make sure that you have the correct number of spaces before <code>input</code>/<code>output</code>/<code>shell</code> and their respective subsections. The number of spaces per level doesn’t matter as long as you’re consistent. Here we use four, but you could just as well use two for a more compact look. Don’t use tabs (unless your editor automatically converts them to spaces).</p>
</div>
</div>
<p>Rules can be given names, here it’s <code>convert_to_upper_case</code>. While rule names are not strictly necessary we encourage you to use them and to make an effort to name your rules in a way that makes it easy to understand the purpose of the rule, as rule names are one of the main ways to interact with the workflow. The <code>shell</code> section (or directive) contains the shell commands that will convert the text in the input file to upper case and send it to the output file. In the shell command string, we can refer to elements of the rule via curly brackets. Here, we refer to the output file by specifying <code>{output}</code> and to the input file by specifying <code>{input}</code>. If you’re not very familiar with Bash, this particular command can be read like “send the contents of <code>a.txt</code> to the program <code>tr</code>, which will convert all characters in the set <code>[a-z]</code> to the corresponding character in the set <code>[A-Z]</code>, and then send the output to <code>a.upper.txt</code>”.</p>
<p>Now let’s run our first Snakemake workflow. When a workflow is executed Snakemake tries to generate a set of target files. Target files can be specified via the command line (or, as you will see later, in several other ways). Here we ask Snakemake to make the file <code>a.upper.txt</code>. We can specify the file containing our rules with <code>-s</code> but since the default behaviour of Snakemake is to look for a file called <code>Snakefile</code> in either the working directory or in a subdirectory called <code>workflow/</code> we don’t need to specify that here. It’s good practice to first run with the flag <code>-n</code> (or <code>--dry-run</code>), which will show what Snakemake plans to do without actually running anything, and you also need to specify how many cores to be used for the workflow with <code>--cores</code> or <code>-c</code>. For now, you only need 1 so set <code>-c 1</code>. You can also use the flag <code>-p</code>, for showing the shell commands that it will execute, and the flag <code>-r</code> for showing the reason for running a specific rule. <code>snakemake --help</code> will show you all available flags.</p>
<pre class="no-highlight"><code>$ snakemake -n -c 1 -r -p a.upper.txt

Building DAG of jobs...
Job stats:
job                      count    min threads    max threads
---------------------  -------  -------------  -------------
convert_to_upper_case        1              1              1
total                        1              1              1


[Mon Oct 25 16:48:43 2021]
rule convert_to_upper_case:
    input: a.txt
    output: a.upper.txt
    jobid: 0
    reason: Missing output files: a.upper.txt
    resources: tmpdir=/var/folders/p0/6z00kpv16qbf_bt52y4zz2kc0000gp/T


        tr [a-z] [A-Z] &lt; a.txt &gt; a.upper.txt

Job stats:
job                      count    min threads    max threads
---------------------  -------  -------------  -------------
convert_to_upper_case        1              1              1
total                        1              1              1

This was a dry-run (flag -n). The order of jobs does not reflect the order of execution.</code></pre>
<p>You can see that Snakemake plans to run one job: the rule <code>convert_to_upper_case</code> with <code>a.txt</code> as input and <code>a.upper.txt</code> as output. The reason for doing this is that it’s missing the file <code>a.upper.txt</code>. Now execute the workflow without the <code>-n</code> flag and check that the contents of <code>a.upper.txt</code> is as expected. Then try running the same command again. What do you see? It turns out that Snakemake only reruns jobs if there have been changes to either <strong>the input files, or the workflow itself</strong>. This is how Snakemake ensures that everything in the workflow is up to date. We will get back to this shortly.</p>
<p>What if we ask Snakemake to generate the file <code>b.upper.txt</code>?</p>
<pre class="no-highlight"><code>$ snakemake -n -c 1 -r -p b.upper.txt

Building DAG of jobs...
MissingRuleException:
No rule to produce b.upper.txt (if you use input functions make sure that they don't raise unexpected exceptions).</code></pre>
<p>That didn’t work well. We could copy the rule to make a similar one for <code>b.txt</code>, but that would be a bit cumbersome. Here is where named wildcards come in; one of the most powerful features of Snakemake. Simply change the input from <code>input: "a.txt"</code> to <code>input: "{some_name}.txt"</code> and the output to <code>output: "{some_name}.upper.txt"</code>. Now try asking for <code>b.upper.txt</code> again.</p>
<p>Tada! What happens here is that Snakemake looks at all the rules it has available (actually only one in this case) and tries to assign values to all wildcards so that the targeted files can be generated. In this case it was quite simple, you can see that it says that <code>wildcards: some_name=b</code>, but for large workflows and multiple wildcards it can get much more complex. Named wildcards is what enables a workflow (or single rules) to be efficiently generalized and reused between projects or shared between people.</p>
<p>It seems we have the first part of our workflow working, now it’s time to make the second rule for concatenating the outputs from <code>convert_to_upper_case</code>. The rule structure will be similar; the only difference is that here we have two inputs instead of one. This can be expressed in two ways, either with named inputs like this:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="bu">input</span>:</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>    firstFile<span class="op">=</span><span class="st">"..."</span>,</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>    secondFile<span class="op">=</span><span class="st">"..."</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>shell:</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""</span></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a><span class="co">    some_function {input.firstFile} {input.secondFile}</span></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Or with indexes like this:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="bu">input</span>:</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>    <span class="co">"..."</span>,</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>    <span class="co">"..."</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>shell:</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a><span class="co">    some_function {input[0]} {input[1]}</span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="callout callout-style-default callout-caution callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Caution
</div>
</div>
<div class="callout-body-container callout-body">
<p>If you have multiple inputs or outputs they need to be delimited with a comma (as seen above). This is a very common mistake when writing Snakemake workflows. The parser will complain, but sometimes the error message can be difficult to interpret.</p>
</div>
</div>
<p>Now try to construct this rule yourself and name it <code>concatenate_a_and_b</code>. The syntax for concatenating two files in Bash is <code>cat first_file.txt second_file.txt &gt; output_file.txt</code>. Call the output <code>c.txt</code>. Run the workflow in Snakemake and validate that the output looks as expected.</p>
<p>Wouldn’t it be nice if our workflow could be used for <em>any</em> files, not just <code>a.txt</code> and <code>b.txt</code>? We can achieve this by using named wildcards (or in other ways as we will discuss later). As we’ve mentioned, Snakemake looks at all the rules it has available and tries to assign values to all wildcards so that the targeted files can be generated. We therefore have to name the output file in a way so that it also contains information about which input files it should be based on. Try to figure out how to do this yourself. If you’re stuck you can look at the spoiler below, but spend some time on it before you look. Also rename the rule to <code>concatenate_files</code> to reflect its new more general use.</p>
<div class="callout callout-style-default callout-tip callout-titled" title="Click to show">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-3-contents" aria-controls="callout-3" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to show
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-3" class="callout-3-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>rule concatenate_files:</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>    output:</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>        <span class="co">"{first}_{second}.txt"</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>    <span class="bu">input</span>:</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>        <span class="co">"{first}.upper.txt"</span>,</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>        <span class="co">"{second}.upper.txt"</span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>    shell:</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""</span></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a><span class="co">        cat {input[0]} {input[1]} &gt; {output}</span></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a><span class="co">        """</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<p>We can now control which input files to use by the name of the file we ask Snakemake to generate. Run the workflow without the flag <code>-n</code> (or <code>--dry-run</code>) to execute both rules, providing one core with <code>-c 1</code> (or <code>--cores 1</code>):</p>
<pre class="no-highlight"><code>$ snakemake a_b.txt -c 1

Building DAG of jobs...
Using shell: /bin/bash
Provided cores: 1 (use --cores to define parallelism)
Rules claiming more threads will be scaled down.
Job stats:
job                      count    min threads    max threads
---------------------  -------  -------------  -------------
concatenate_files            1              1              1
convert_to_upper_case        2              1              1
total                        3              1              1

Select jobs to execute...

[Mon Oct 25 16:51:52 2021]
rule convert_to_upper_case:
    input: b.txt
    output: b.upper.txt
    jobid: 2
    wildcards: some_name=b
    resources: tmpdir=/var/folders/p0/6z00kpv16qbf_bt52y4zz2kc0000gp/T

[Mon Oct 25 16:51:53 2021]
Finished job 2.
1 of 3 steps (33%) done
Select jobs to execute...

[Mon Oct 25 16:51:53 2021]
rule convert_to_upper_case:
    input: a.txt
    output: a.upper.txt
    jobid: 1
    wildcards: some_name=a
    resources: tmpdir=/var/folders/p0/6z00kpv16qbf_bt52y4zz2kc0000gp/T

[Mon Oct 25 16:51:53 2021]
Finished job 1.
2 of 3 steps (67%) done
Select jobs to execute...

[Mon Oct 25 16:51:53 2021]
rule concatenate_files:
    input: a.upper.txt, b.upper.txt
    output: a_b.txt
    jobid: 0
    wildcards: first=a, second=b
    resources: tmpdir=/var/folders/p0/6z00kpv16qbf_bt52y4zz2kc0000gp/T

[Mon Oct 25 16:51:53 2021]
Finished job 0.
3 of 3 steps (100%) done</code></pre>
<p>Neat!</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>You can name a file whatever you want in a Snakemake workflow, but you will find that everything falls into place much nicer if the filename reflects the file’s path through the workflow, <em>e.g.</em> <code>sample_a.trimmed.deduplicated.sorted.bam</code>.</p>
</div>
</div>
<p>The input to Snakemake rules have to be strings or lists of strings, however you don’t have to specify these strings directly in the <code>input:</code> section of rules. Instead, you can specify Python functions that return strings or lists of strings. This allows you to supply input to rules that can vary depending on the wildcards being used. We’ll get to why that’s useful in a sec, but first let’s put it to use for the <code>conatenate_files</code> rule. Because Snakemake is based on Python we can mix rule definitions with standard python code in the same file. Add a function just above the <code>concatenate_files</code> that looks like this:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> concat_input(wildcards):</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>    files <span class="op">=</span> [wildcards.first <span class="op">+</span> <span class="st">".upper.txt"</span>, wildcards.second <span class="op">+</span> <span class="st">".upper.txt"</span>]</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> files</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>This is the syntax to define a function in Python. The <code>def concat_input(wildcards):</code> line shows the name of the function (<code>concat_input</code>) and the variable passed to the function (the <code>wildcards</code> object). In the second line we add two items to a list that we call <code>files</code> and add the ‘.upper.txt’ suffix to each item. Finally, the function returns the list. Because the <code>concatenate_files</code> rule has two wildcards <code>{first}</code> and <code>{second}</code> we can access the actual strings in the <code>wildcards</code> object using <code>wildcards.first</code> and <code>wildcards.second</code>. When we ask for the file <code>a_b.txt</code> then <code>wildcards.first == 'a'</code> and <code>wildcards.second == 'b'</code>. This means that the <code>files</code> list returned by the function will be <code>['a.upper.txt', 'b.upper.txt']</code>. To see for yourself you can add the following line to the function, just before the return statement: <code>print (wildcards.first, wildcards.second, files)</code>. This way the wildcard values and the list will be printed to the terminal when you run Snakemake.</p>
<p>Now that we’ve defined the function to use as input, we can use it in the <code>concatenate_files</code> rule. Update the rule so that it looks like this:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>rule concatenate_files:</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>    output:</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>        <span class="co">"{first}_{second}.txt"</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>    <span class="bu">input</span>:</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>        concat_input</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>    shell:</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a><span class="co">        cat {input[0]} {input[1]} &gt; {output}</span></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a><span class="co">        """</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>You see that the name of the function <code>concat_input</code> is added in place of the input strings. When using the <code>wildcards</code> object in input functions like this we have to call the function without any arguments (simply <code>concat_input</code>) and the function has to be defined to accept a single argument (here <code>def concat_input(wildcards):</code>). Let’s run the workflow with the updated rule. Remove the file <code>a_b.txt</code> or add <code>-f</code> to the Snakemake command to force a re-run:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="ex">snakemake</span> a_b.txt <span class="at">-c</span> 1 <span class="at">-f</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>If you added the <code>print</code> statement to the function you should see the following printed to your terminal:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="ex">Building</span> DAG of jobs...</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="ex">a</span> b [<span class="st">'a.upper.txt'</span>, <span class="st">'b.upper.txt'</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Followed by the rest of the workflow output.</p>
<p>There are a number of possible use-cases for input functions. For example, say that you have an experiment where you’ve sequenced three samples: <code>sample1</code>, <code>sample2</code> and <code>sample3</code> with the corresponding FASTQ files under <code>data/</code> and you want to write a rule that outputs the statistics of all sequences within each sample. However, samples <code>sample1</code> and <code>sample2</code> have been sequenced with single-end technology while <code>sample3</code> have paired-end reads. The single-end samples will have only one FASTQ file whereas the paired-end sample will have two (one for each sequenced end). Thus, depending on the name of the sample the input to the function will either be one file or two. With input functions we can write a generalized rule that can handle both types:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> fastq_input(wildcards):</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> wildcards.sample_id <span class="kw">in</span> [<span class="st">"sample1"</span>, <span class="st">"sample2"</span>]:</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> <span class="st">"data/"</span> <span class="op">+</span> wildcards.sample_id <span class="op">+</span> <span class="st">".fastq.gz"</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> [<span class="st">"data/"</span> <span class="op">+</span> wildcards.sample_id <span class="op">+</span> <span class="st">".R1.fastq.gz"</span>,</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>                <span class="st">"data/"</span> <span class="op">+</span> wildcards.sample_id <span class="op">+</span> <span class="st">".R2.fastq.gz"</span>]</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>rule fastq_stats:</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>    output:</span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>        <span class="co">"{sample_id}.stats.txt"</span></span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>    <span class="bu">input</span>:</span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a>        fastq_input</span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a>    shell:</span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""</span></span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a><span class="co">        seqtk comp {input} &gt; {output}</span></span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a><span class="co">        """</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>As you can see, the <code>fastq_stats</code> rule outputs one file <code>{sample_id}.stats.txt</code> and takes as input the value returned from the <code>fastq_input</code> function. In this function the sample id is evaluated and if it is either <code>sample1</code> or <code>sample2</code> (our single-end samples) then the function returns a single string which is the path to the FASTQ file for that sample. Otherwise, the function returns a list containing both the <code>R1</code> and <code>R2</code> files for the sample. In the <code>shell:</code> directive of the rule the <code>seqtk comp</code> command is run on the input and the output is sent to the output file.</p>
<div class="callout callout-style-default callout-note callout-titled" title="Quick recap">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Quick recap
</div>
</div>
<div class="callout-body-container callout-body">
<p>In this section we’ve learned:</p>
<ul>
<li>How a simple Snakemake rule looks.</li>
<li>How to define target files when executing a workflow.</li>
<li>How to use named wildcards for writing generic and flexible rules.</li>
<li>How to use input functions in rules</li>
</ul>
</div>
</div>
</section>
<section id="visualising-workflows" class="level2" data-number="3">
<h2 data-number="3" class="anchored" data-anchor-id="visualising-workflows"><span class="header-section-number">3</span> Visualising workflows</h2>
<p>All that we’ve done so far could quite easily be done in a simple shell script that takes the input files as parameters. Let’s now take a look at some of the features where a WfMS like Snakemake really adds value compared to a more straightforward approach. One such feature is the possibility to visualize your workflow. Snakemake can generate three types of graphs, one that shows how the rules are connected, one that shows how the jobs (<em>i.e.</em> an execution of a rule with some given inputs/outputs/settings) are connected, and finally one that shows rules with their respective input/output files.</p>
<p>First we look at the rule graph. The following command will generate a rule graph in the dot language and pipe it to the program <code>dot</code>, which in turn will save a visualization of the graph as a PNG file (if you’re having troubles displaying PNG files you could use SVG or JPG instead).</p>
<div class="callout callout-style-default callout-caution callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Caution
</div>
</div>
<div class="callout-body-container callout-body">
<p>If you added the <code>print(wildcards.first,wildcards.second,files)</code> statement to the <code>concat_input</code> function in the previous section you need to remove that line before running the commands below.</p>
</div>
</div>
<div class="sourceCode" id="cb14"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="ex">snakemake</span> <span class="at">--rulegraph</span> a_b.txt <span class="kw">|</span> <span class="ex">dot</span> <span class="at">-Tpng</span> <span class="op">&gt;</span> rulegraph.png</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><img src="images/rulegraph.svg" class="img-fluid"></p>
<p>This looks simple enough, the output from the rule <code>convert_to_upper_case</code> will be used as input to the rule <code>concatenate_files</code>.</p>
<p>For a more typical bioinformatics project it can look something like this when you include all the rules from processing of the raw data to generating figures for the paper.</p>
<p><img src="images/rulegraph_complex.svg" class="img-fluid"></p>
<p>While saying that it’s easy to read might be a bit of a stretch, it definitely gives you a better overview of the project than you would have without a WfMS.</p>
<p>The second type of graph is based on the jobs, and looks like this for our little workflow (use <code>--dag</code> instead of <code>--rulegraph</code>).</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="ex">snakemake</span> <span class="at">--dag</span> a_b.txt <span class="kw">|</span> <span class="ex">dot</span> <span class="at">-Tpng</span> <span class="op">&gt;</span> jobgraph.png</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><img src="images/jobgraph.svg" class="img-fluid"></p>
<p>The main difference here is that now each node is a job instead of a rule. You can see that the wildcards used in each job are also displayed. Another difference is the dotted lines around the nodes. A dotted line is Snakemake’s way of indicating that this rule doesn’t need to be rerun in order to generate <code>a_b.txt</code>. Validate this by running <code>snakemake -n -r a_b.txt</code> and it should say that there is nothing to be done.</p>
<p>We’ve discussed before that one of the main purposes of using a WfMS is that it automatically makes sure that everything is up to date. This is done by recursively checking that outputs are always newer than inputs for all the rules involved in the generation of your target files. Now try to change the contents of <code>a.txt</code> to some other text and save it. What do you think will happen if you run <code>snakemake -n -r a_b.txt</code> again?</p>
<div class="callout callout-style-default callout-tip callout-titled" title="Click to show">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-7-contents" aria-controls="callout-7" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to show
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-7" class="callout-7-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<pre class="no-highlight"><code>$ snakemake -n -r a_b.txt

Building DAG of jobs...
Job stats:
job                      count    min threads    max threads
---------------------  -------  -------------  -------------
concatenate_files            1              1              1
convert_to_upper_case        1              1              1
total                        2              1              1


[Mon Oct 25 17:00:02 2021]
rule convert_to_upper_case:
    input: a.txt
    output: a.upper.txt
    jobid: 1
    reason: Updated input files: a.txt
    wildcards: some_name=a
    resources: tmpdir=/var/folders/p0/6z00kpv16qbf_bt52y4zz2kc0000gp/T


[Mon Oct 25 17:00:02 2021]
rule concatenate_files:
    input: a.upper.txt, b.upper.txt
    output: a_b.txt
    jobid: 0
    reason: Input files updated by another job: a.upper.txt
    wildcards: first=a, second=b
    resources: tmpdir=/var/folders/p0/6z00kpv16qbf_bt52y4zz2kc0000gp/T

Job stats:
job                      count    min threads    max threads
---------------------  -------  -------------  -------------
concatenate_files            1              1              1
convert_to_upper_case        1              1              1
total                        2              1              1

This was a dry-run (flag -n). The order of jobs does not reflect the order of execution.</code></pre>
</div>
</div>
</div>
<p>Were you correct? Also generate the job graph and compare to the one generated above. What’s the difference? Now rerun without <code>-n</code> and validate that <code>a_b.txt</code> contains the new text (don’t forget to specify <code>-c 1</code>). Note that Snakemake doesn’t look at the contents of files when trying to determine what has changed, only at the timestamp for when they were last modified.</p>
<p>We’ve seen that Snakemake keeps track of if files in the workflow have changed, and automatically makes sure that any results depending on such files are regenerated. What about if the rules themselves are changed? It turns out that since version 7.8.0 Snakemake keeps track of this automatically.</p>
<p>Let’s say that we want to modify the rule <code>concatenate_files</code> to also include which files were concatenated.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>rule concatenate_files:</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>    output:</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>        <span class="co">"{first}_{second}.txt"</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>    <span class="bu">input</span>:</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>        <span class="co">"{first}.upper.txt"</span>,</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>        <span class="co">"{second}.upper.txt"</span></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>    shell:</span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""</span></span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a><span class="co">        echo 'Concatenating {input}' | cat - {input[0]} {input[1]} &gt; {output}</span></span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a><span class="co">        """</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>It’s not really important for the tutorial, but the shell command used here first outputs “Concatenating” followed by a space delimited list of the files in <code>input</code>. This string is then sent to the program <code>cat</code> where it’s concatenated with <code>input[0]</code> and <code>input[1]</code> (the parameter <code>-</code> means that it should read from standard input). Lastly, the output from <code>cat</code> is sent to <code>{output}</code>.</p>
</div>
</div>
<p>If you now run the workflow as before you should see:</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="ex">rule</span> concatenate_files:</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>    <span class="ex">input:</span> a.upper.txt, b.upper.txt</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>    <span class="ex">output:</span> a_b.txt</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>    <span class="ex">jobid:</span> 0</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>    <span class="ex">reason:</span> Code has changed since last execution</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>    <span class="ex">wildcards:</span> first=a, second=b</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Because although no files involved in the workflow have been changed, Snakemake recognizes that the workflow code itself has been modified and this triggers a re-run.</p>
<p>Snakemake is aware of changes to four categories of such “rerun-triggers”: “input” (changes to rule input files), “params” (changes to the rule <code>params</code> section), “software-env” (changes to Conda environment files specified by the <code>conda:</code> directive) and “code” (changes to code in the <code>shell:</code>, <code>run:</code>, <code>script:</code> and <code>notebook:</code> directives).</p>
<p>Prior to version 7.8.0, only changes to the modification time of input files would trigger automatic re-runs. To run Snakemake with this previous behaviour you can use the setting <code>--rerun-triggers mtime</code> at the command line. Change the <code>shell:</code> section of the <code>concatenate_files</code> rule back to the previous version, then try running: <code>snakemake -n -r a_b.txt --rerun-triggers mtime</code> and you should again see <code>Nothing to be done (all requested files are present and up to date).</code></p>
<p>You can also export information on how all files were generated (when, by which rule, which version of the rule, and by which commands) to a tab-delimited file like this:</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="ex">snakemake</span> a_b.txt <span class="at">-c</span> 1 <span class="at">-D</span> <span class="op">&gt;</span> summary.tsv</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>The content of <code>summary.tsv</code> is shown in the table below:</p>
<div class="table-responsive">
<table class="table">
<colgroup>
<col style="width: 7%">
<col style="width: 13%">
<col style="width: 11%">
<col style="width: 5%">
<col style="width: 7%">
<col style="width: 12%">
<col style="width: 18%">
<col style="width: 14%">
<col style="width: 8%">
</colgroup>
<thead>
<tr class="header">
<th>output_file</th>
<th>date</th>
<th>rule</th>
<th>version</th>
<th>log-file(s)</th>
<th>input-file(s)</th>
<th>shellcmd</th>
<th>status</th>
<th>plan</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>a_b.txt</td>
<td>Mon Oct 25 17:01:46 2021</td>
<td>concatenate_files</td>
<td>-</td>
<td></td>
<td>a.upper.txt,b.upper.txt</td>
<td>cat a.upper.txt b.upper.txt &gt; a_b.txt</td>
<td>rule implementation changed</td>
<td>update pending</td>
</tr>
<tr class="even">
<td>a.upper.txt</td>
<td>Mon Oct 25 17:01:46 2021</td>
<td>convert_to_upper_case</td>
<td>-</td>
<td></td>
<td>a.txt</td>
<td>tr [a-z] [A-Z] &lt; a.txt &gt; a.upper.txt</td>
<td>ok</td>
<td>no update</td>
</tr>
<tr class="odd">
<td>b.upper.txt</td>
<td>Mon Oct 25 17:01:46 2021</td>
<td>convert_to_upper_case</td>
<td>-</td>
<td></td>
<td>b.txt</td>
<td>tr [a-z] [A-Z] &lt; b.txt &gt; b.upper.txt</td>
<td>ok</td>
<td>no update</td>
</tr>
</tbody>
</table>
</div>
<p>You can see in the second last column that the rule implementation for <code>a_b.txt</code> has changed. The last column shows if Snakemake plans to regenerate the files when it’s next executed. You can see that for the <code>concatenate_files</code> the plan is <code>update pending</code> because we generated the summary with the default behaviour of using all rerun-triggers.</p>
<p>You might wonder where Snakemake keeps track of all these things? It stores all information in a hidden subdirectory called <code>.snakemake</code>. This is convenient since it’s easy to delete if you don’t need it anymore and everything is contained in the project directory. Just be sure to add it to <code>.gitignore</code> so that you don’t end up tracking it with git.</p>
<p>By now you should be familiar with the basic functionality of Snakemake, and you can build advanced workflows with only the features we have discussed here. There’s a lot we haven’t covered though, in particular when it comes to making your workflow more reusable. In the following section we will start with a workflow that is fully functional but not very flexible. We will then gradually improve it, and at the same time showcase some Snakemake features we haven’t discussed yet. Note that this can get a little complex at times, so if you felt that this section was a struggle then you could move on to one of the other tutorials instead.</p>
<div class="callout callout-style-default callout-note callout-titled" title="Quick recap">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Quick recap
</div>
</div>
<div class="callout-body-container callout-body">
<p>In this section we’ve learned:</p>
<ul>
<li>How to use <code>--dag</code> and <code>--rulegraph</code> for visualizing the job and rule graphs, respectively.</li>
<li>How Snakemake reruns relevant parts of the workflow after there have been changes.</li>
<li>How Snakemake tracks changes to files and code in a workflow</li>
</ul>
</div>
</div>
</section>
<section id="the-mrsa-workflow" class="level2" data-number="4">
<h2 data-number="4" class="anchored" data-anchor-id="the-mrsa-workflow"><span class="header-section-number">4</span> The MRSA workflow</h2>
<p>As you might remember from the <a href="introduction">intro</a>, we are attempting to understand how lytic bacteriophages can be used as a future therapy for the multi-resistant bacteria MRSA (methicillin-resistant <em>Staphylococcus aureus</em>). In order to do this we have performed RNA-seq of three strains, one test and two controls. We have already set up a draft Snakemake workflow for the RNA-seq analysis and it seems to be running nicely. The rest of the Snakemake tutorial will be spent improving and making this workflow more flexible!</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>This section will leave a little more up to you compared to the previous one. If you get stuck at some point the final workflow after all the modifications is available in <code>tutorials/git/Snakefile</code>.</p>
</div>
</div>
<p>You are probably already in your <code>snakemake-env</code> environment, otherwise activate it (use <code>conda info --envs</code> if you are unsure).</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>Here we have one Conda environment for executing the whole Snakemake workflow. Snakemake also supports using explicit Conda environments on a per-rule basis, by specifying something like <code>conda: rule-specific-env.yml</code> in the rule definition and running Snakemake with the <code>--use-conda</code> flag. The given rule will then be run in the Conda environment specified in <code>rule-specific-env.yml</code> that will be created and activated on the fly by Snakemake. Note that by default Snakemake uses <code>conda</code> to generate the rule-specific environments. This behaviour can be changed by running with <code>--conda-frontend conda</code>, which will force Snakemake to use <code>conda</code> instead.</p>
</div>
</div>
<p>Let’s start by generating the rule graph so that we get an overview of the workflow. Here we have to specify the file with the rules using the <code>-s</code> flag to Snakemake since the path to the file differs from the default.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="ex">snakemake</span> <span class="at">-s</span> snakefile_mrsa.smk <span class="at">--rulegraph</span> <span class="kw">|</span> <span class="ex">dot</span> <span class="at">-T</span> png <span class="op">&gt;</span> rulegraph_mrsa.png</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>There’s another difference in this command compared to the one we’ve used before, namely that we don’t define a target. In the toy example we used <code>a_b.txt</code> as a target, and the wildcards were resolved based on that. How come that we don’t need to do that here? It turns out that by default Snakemake targets the first rule in a workflow. By convention, we call this rule <code>all</code> and let it serve as a rule for aggregating the main outputs of the workflow.</p>
<p><img src="images/rulegraph_mrsa.svg" class="img-fluid"></p>
<p>Now take some time and look through the workflow file and try to understand how the rules fit together. Use the rule graph as aid. The rules represent a quite standard, although somewhat simplified, workflow for RNA-seq analysis. If you are unfamiliar with the purpose of the different operations (index genome, FastQC and so on), then take a look at the <a href="introduction">intro</a>.</p>
<p>Also generate the job graph in the same manner. Here you can see that three samples will be downloaded: SRR935090, SRR935091, and SRR935092. The original sample files contain tens of millions of reads but for the purpose of this course we have sub-sampled them to 100,000 reads per sample, so that they are easy to manage, and made them available at the <a href="https://figshare.scilifelab.se/articles/educational_resource/MRSA_case_study_example_data/22246417">SciLifeLab Data Repository</a>. These FASTQ files will then be quality controlled with FastQC and aligned to a genome. The QC output will be aggregated with MultiQC and the alignments will be used to generate a count table, <em>i.e.</em> a table that shows how many reads map to each gene for each sample. This count table is then what the downstream analysis will be based on.</p>
<p><img src="images/dag_mrsa.svg" class="img-fluid"></p>
<p>Now try to run the whole workflow. Hopefully you see something like this.</p>
<pre class="no-highlight"><code>Building DAG of jobs...
Using shell: /bin/bash
Provided cores: 1 (use --cores to define parallelism)
Rules claiming more threads will be scaled down.
Job stats:
job                     count    min threads    max threads
--------------------  -------  -------------  -------------
align_to_genome             3              1              1
all                         1              1              1
fastqc                      3              1              1
generate_count_table        1              1              1
generate_rulegraph          1              1              1
get_SRA_by_accession        3              1              1
get_genome_fasta            1              1              1
get_genome_gff3             1              1              1
index_genome                1              1              1
multiqc                     1              1              1
sort_bam                    3              1              1
total                      19              1              1

Select jobs to execute...

[Mon Oct 25 17:13:47 2021]
rule get_genome_fasta:
    output: data/ref/NCTC8325.fa.gz
    jobid: 6
    resources: tmpdir=/var/folders/p0/6z00kpv16qbf_bt52y4zz2kc0000gp/T

--2021-10-25 17:13:48--  ftp://ftp.ensemblgenomes.org/pub/bacteria/release-37/fasta/bacteria_18_collection/staphylococcus_aureus_subsp_aureus_nctc_8325/dna//Staphylococcus_aureus_subsp_aureus_nctc_8325.ASM1342v1.dna_rm.toplevel.fa.gz
           =&gt; ‘data/ref/NCTC8325.fa.gz’
Resolving ftp.ensemblgenomes.org (ftp.ensemblgenomes.org)... 193.62.197.75
Connecting to ftp.ensemblgenomes.org (ftp.ensemblgenomes.org)|193.62.197.75|:21... connected.
Logging in as anonymous ... Logged in!
==&gt; SYST ... done.    ==&gt; PWD ... done.
.
.
[lots of stuff]
.
.
localrule all:
    input: results/tables/counts.tsv, results/multiqc/multiqc.html, results/rulegraph.png
    jobid: 0
    resources: tmpdir=/var/folders/p0/6z00kpv16qbf_bt52y4zz2kc0000gp/T

[Mon Oct 25 17:14:38 2021]
Finished job 0.
19 of 19 steps (100%) done</code></pre>
<p>After everything is done, the workflow will have resulted in a bunch of files in the directories <code>data/</code> and <code>results/</code>. Take some time to look through the structure, in particular the quality control reports in <code>results/multiqc/</code> and the count table in <code>results/tables/</code>.</p>
<div class="callout callout-style-default callout-note callout-titled" title="Quick recap">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Quick recap
</div>
</div>
<div class="callout-body-container callout-body">
<p>In this section we’ve learned:</p>
<ul>
<li>How the MRSA workflow looks.</li>
<li>How to run the MRSA workflow.</li>
<li>Which output files the MRSA workflow produces.</li>
</ul>
</div>
</div>
</section>
<section id="parameters" class="level2" data-number="5">
<h2 data-number="5" class="anchored" data-anchor-id="parameters"><span class="header-section-number">5</span> Parameters</h2>
<p>In a typical bioinformatics project, considerable efforts are spent on tweaking parameters for the various programs involved. It would be inconvenient if you had to change in the shell scripts themselves every time you wanted to run with a new setting. Luckily, there is a better option for this: the <code>params</code> keyword.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>rule some_rule:</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>    output:</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>        <span class="co">"..."</span></span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>    <span class="bu">input</span>:</span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>        <span class="co">"..."</span></span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>    params:</span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a>        cutoff<span class="op">=</span><span class="fl">2.5</span></span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a>    shell:</span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""</span></span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a><span class="co">        some_program --cutoff {params.cutoff} {input} {output}</span></span>
<span id="cb22-11"><a href="#cb22-11" aria-hidden="true" tabindex="-1"></a><span class="co">        """</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Most of the programs are run with default settings in the MRSA workflow and don’t use the <code>params:</code> directive. However, the <code>get_SRA_by_accession</code> rule is an exception. Here the remote address for each of the files to download is passed to the shell directive via:</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_sample_url(wildcards):</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>    samples <span class="op">=</span> {</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>        <span class="st">"SRR935090"</span>: <span class="st">"https://figshare.scilifelab.se/ndownloader/files/39539767"</span>,</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>        <span class="st">"SRR935091"</span>: <span class="st">"https://figshare.scilifelab.se/ndownloader/files/39539770"</span>,</span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>        <span class="st">"SRR935092"</span>: <span class="st">"https://figshare.scilifelab.se/ndownloader/files/39539773"</span></span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> samples[wildcards.sample_id]</span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a>rule get_SRA_by_accession:</span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""</span></span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a><span class="co">    Retrieve a single-read FASTQ file</span></span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true" tabindex="-1"></a>    output:</span>
<span id="cb23-14"><a href="#cb23-14" aria-hidden="true" tabindex="-1"></a>        <span class="co">"data/{sample_id}.fastq.gz"</span></span>
<span id="cb23-15"><a href="#cb23-15" aria-hidden="true" tabindex="-1"></a>    params:</span>
<span id="cb23-16"><a href="#cb23-16" aria-hidden="true" tabindex="-1"></a>        url <span class="op">=</span> get_sample_url</span>
<span id="cb23-17"><a href="#cb23-17" aria-hidden="true" tabindex="-1"></a>    shell:</span>
<span id="cb23-18"><a href="#cb23-18" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""</span></span>
<span id="cb23-19"><a href="#cb23-19" aria-hidden="true" tabindex="-1"></a><span class="co">        wget -O - {params.url} | seqtk sample - 25000 | gzip -c &gt; {output[0]}</span></span>
<span id="cb23-20"><a href="#cb23-20" aria-hidden="true" tabindex="-1"></a><span class="co">        """</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>You may recognize this from <a href="snakemake-2-the-basics">page 2</a> of this tutorial where we used input functions to generate strings and lists of strings for the <code>input:</code> section of a rule. Using a function to return values based on the wildcards also works for <code>params:</code>. Here <code>sample_id</code> is a wildcard which in this specific workflow can be either <code>SRR935090</code>, <code>SRR935091</code>, or <code>SRR935092</code>. The wildcards object is passed to the function <code>get_sample_url</code> and depending on what output the rule is supposed to generate, <code>wildcards.sample_id</code> will take the value of either of the three sample ids. The <code>samples</code> variable defined in the function is a Python <a href="https://docs.python.org/3/tutorial/datastructures.html#dictionaries">dictionary</a> that has the URLs for each sample_id hard-coded. This dictionary is used to convert the value of the <code>sample_id</code> wildcard to a URL, which is returned by the function. Finally, in the <code>shell:</code> directive we access the <code>url</code> parameter with <code>{params.url}</code>. (We could have written three separate rules to download the samples, but it’s easy to see how that can become impractical.)</p>
<p>Let’s add another parameter to the <code>get_SRA_by_accession</code> rule. As you can see in the shell command the FASTQ file downloaded by <code>wget</code> gets piped directly (the <code>-O -</code> part means send contents to STDOUT) to the <code>seqtk sample</code> command which reads from STDIN and outputs 25000 randomly sampled reads (out of the 100,000 contained in the example FASTQ file). Change in the rule to use the parameter <code>max_reads</code> instead and set the value to 20000. If you need help, click to show the solution below.</p>
<div class="callout callout-style-default callout-tip callout-titled" title="Click to show">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-13-contents" aria-controls="callout-13" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to show
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-13" class="callout-13-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb24"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>rule get_SRA_by_accession:</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""</span></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a><span class="co">    Retrieve a single-read FASTQ file</span></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>    output:</span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>        <span class="co">"data/{sample_id}.fastq.gz"</span></span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>    params:</span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>        url <span class="op">=</span> get_sample_url,</span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a>        max_reads <span class="op">=</span> <span class="dv">20000</span></span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a>    shell:</span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""</span></span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a><span class="co">        wget -O - {params.url} | seqtk sample - {params.max_reads} | gzip -c &gt; {output[0]}</span></span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a><span class="co">        """</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<p>Now run through the workflow. Because there’s been changes to the <code>get_SRA_by_accession</code> rule this will trigger a re-run of the rule for all three accessions. In addition all downstream rules that depend on output from <code>get_SRA_by_accession</code> are re-run.</p>
<p>As you can see the parameter values we set in the <code>params</code> section don’t have to be static, they can be any Python expression. In particular, Snakemake provides a global dictionary of configuration parameters called <code>config</code>. Let’s modify <code>get_SRA_by_accession</code> to look something like this in order to make use of this dictionary:</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>rule get_SRA_by_accession:</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""</span></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a><span class="co">    Retrieve a single-read FASTQ file</span></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>    output:</span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>        <span class="co">"data/{sample_id}.fastq.gz"</span></span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>    params:</span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>        url <span class="op">=</span> get_sample_url,</span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a>        max_reads <span class="op">=</span> config[<span class="st">"max_reads"</span>]</span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a>    shell:</span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""</span></span>
<span id="cb25-12"><a href="#cb25-12" aria-hidden="true" tabindex="-1"></a><span class="co">        wget -L {params.url} | seqtk sample - {params.max_reads} | gzip -c &gt; {output[0]}</span></span>
<span id="cb25-13"><a href="#cb25-13" aria-hidden="true" tabindex="-1"></a><span class="co">        """</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Note that Snakemake now expects there to be a key named <code>max_reads</code> in the config dictionary. If we don’t populate the dictionary somehow the dictionary will be empty so if you were to run the workflow now it would trigger a <code>KeyError</code> (try running <code>snakemake -s snakefile_mrsa.smk -n</code> to see for yourself). In order to populate the config dictionary with data for the workflow we could use the <code>snakemake --config KEY=VALUE</code> syntax directly from the command line (<em>e.g.</em> <code>snakemake --config max_reads=20000 -s snakefile_mrsa.smk</code>). However, from a reproducibility perspective, it’s not optimal to set parameters from the command line, since it’s difficult to keep track of which parameter values that were used.</p>
<p>A much better alternative is to use the <code>--configfile FILE</code> option to supply a configuration file to Snakemake. In this file we can collect all the project-specific settings, sample ids and so on. This also enables us to write the Snakefile in a more general manner so that it can be better reused between projects. Like several other files used in these tutorials, this file should be in <a href="https://en.wikipedia.org/wiki/YAML">YAML format</a>. Create the file below and save it as <code>config.yml</code>.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode yaml code-with-copy"><code class="sourceCode yaml"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">max_reads</span><span class="kw">:</span><span class="at"> </span><span class="dv">25000</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>If we now run Snakemake with <code>--configfile config.yml</code>, it will parse this file to form the <code>config</code> dictionary. If you want to overwrite a parameter value, <em>e.g.</em> for testing, you can still use the <code>--config KEY=VALUE</code> flag, as in <code>--config max_reads=1000</code>.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>Rather than supplying the config file from the command line you could also add the line <code>configfile: "config.yml"</code> to the top of your Snakefile. Keep in mind that with such a setup Snakemake will complain if the file <code>config.yml</code> is not present.</p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled" title="Quick recap">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Quick recap
</div>
</div>
<div class="callout-body-container callout-body">
<p>In this section we’ve learned:</p>
<ul>
<li>How to set parameter values with the <code>params</code> directive.</li>
<li>How to run Snakemake with the <code>config</code> variable and with a configuration file.</li>
</ul>
</div>
</div>
</section>
<section id="logs" class="level2" data-number="6">
<h2 data-number="6" class="anchored" data-anchor-id="logs"><span class="header-section-number">6</span> Logs</h2>
<p>As you probably noticed it was difficult to follow how the workflow progressed since some rules printed a lot of output to the terminal. In some cases this also contained important information, such as statistics on the sequence alignments or genome indexing. This could be valuable for example if you later in the project get weird results and want to debug. It’s also important from a reproducibility perspective that the “paper trail” describing how the outputs were generated is saved. Luckily, Snakemake has a feature that can help with this. Just as we define <code>input</code> and <code>output</code> in a rule we can also define <code>log</code>.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>rule some_rule:</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>    output:</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>        <span class="co">"..."</span></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>    <span class="bu">input</span>:</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>        <span class="co">"..."</span></span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>    log:</span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a>        <span class="co">"..."</span></span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a>    shell:</span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""</span></span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a><span class="co">        echo 'Converting {input} to {output}' &gt; {log}</span></span>
<span id="cb27-11"><a href="#cb27-11" aria-hidden="true" tabindex="-1"></a><span class="co">        """</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>A log file is not different from any other output file, but it’s dealt with a little differently by Snakemake. For example, it’s shown in the file summary when using <code>-D</code> and unlike other output files it’s not deleted if jobs fail which of course is necessary for debugging purposes. It’s also a good way to clarify the purpose of the file. We probably don’t need to save logs for all the rules, only the ones with interesting output.</p>
<ul>
<li><code>get_genome_fasta</code> and <code>get_genome_gff3</code> would be good to log since they are dependent on downloading files from an external server.</li>
<li><code>multiqc</code> aggregates quality control data for all the samples into one html report, and the log contains information about which samples were aggregated.</li>
<li><code>index_genome</code> outputs some statistics about the genome indexing.</li>
<li><code>align_to_genome</code> outputs important statistics about the alignments. This is probably the most important log to save.</li>
</ul>
<p>Now add a log file to some or all of the rules above. A good place to save them to would be <code>results/logs/rule_name/</code>. In order to avoid that multiple jobs write to the same files Snakemake requires that all output and log files contain the same wildcards, so be sure to include any wildcards used in the rule in the log name as well, <em>e.g.</em> <code>{some_wildcard}.log</code>.</p>
<p>You also have to specify in the <code>shell</code> section of each rule what you want the log to contain. Some of the programs we use send their log information to standard out, some to standard error and some let us specify a log file via a flag.</p>
<p>For example, in the <code>align_to_genome</code> rule, it could look like this (Bowtie2 writes log info to standard error):</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>rule align_to_genome:</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""</span></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a><span class="co">    Align a fastq file to a genome index using Bowtie 2.</span></span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>    output:</span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a>        <span class="co">"results/bam/{sample_id,\w+}.bam"</span></span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a>    <span class="bu">input</span>:</span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a>        <span class="co">"data/{sample_id}.fastq.gz"</span>,</span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a>        <span class="co">"results/bowtie2/NCTC8325.1.bt2"</span>,</span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a>        <span class="co">"results/bowtie2/NCTC8325.2.bt2"</span>,</span>
<span id="cb28-11"><a href="#cb28-11" aria-hidden="true" tabindex="-1"></a>        <span class="co">"results/bowtie2/NCTC8325.3.bt2"</span>,</span>
<span id="cb28-12"><a href="#cb28-12" aria-hidden="true" tabindex="-1"></a>        <span class="co">"results/bowtie2/NCTC8325.4.bt2"</span>,</span>
<span id="cb28-13"><a href="#cb28-13" aria-hidden="true" tabindex="-1"></a>        <span class="co">"results/bowtie2/NCTC8325.rev.1.bt2"</span>,</span>
<span id="cb28-14"><a href="#cb28-14" aria-hidden="true" tabindex="-1"></a>        <span class="co">"results/bowtie2/NCTC8325.rev.2.bt2"</span></span>
<span id="cb28-15"><a href="#cb28-15" aria-hidden="true" tabindex="-1"></a>    log:</span>
<span id="cb28-16"><a href="#cb28-16" aria-hidden="true" tabindex="-1"></a>        <span class="co">"results/logs/align_to_genome/{sample_id}.log"</span></span>
<span id="cb28-17"><a href="#cb28-17" aria-hidden="true" tabindex="-1"></a>    shell:</span>
<span id="cb28-18"><a href="#cb28-18" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""</span></span>
<span id="cb28-19"><a href="#cb28-19" aria-hidden="true" tabindex="-1"></a><span class="co">        bowtie2 -x results/bowtie2/NCTC8325 -U {input[0]} &gt; {output} 2&gt;{log}</span></span>
<span id="cb28-20"><a href="#cb28-20" aria-hidden="true" tabindex="-1"></a><span class="co">        """</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>To save some time you can use the info below.</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co"># wget has a -o flag for specifying the log file</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> remote_file <span class="at">-O</span> output_file <span class="at">-o</span> {log}</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a><span class="co"># MultiQC and featureCounts write to standard error so we redirect with "2&gt;"</span></span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a><span class="ex">multiqc</span> <span class="at">-n</span> output_file input_files <span class="dv">2</span><span class="op">&gt;</span> {log}</span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a><span class="ex">featureCounts</span> <span class="at">-t</span> gene <span class="at">-g</span> gene_id <span class="at">-a</span> gff_file <span class="at">-o</span> output_file input_files <span class="dv">2</span><span class="op">&gt;</span>{log}</span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Bowtie2-build redirects to standard out so we use "&gt;"</span></span>
<span id="cb29-9"><a href="#cb29-9" aria-hidden="true" tabindex="-1"></a><span class="ex">bowtie2-build</span> input_file index_dir <span class="op">&gt;</span> {log}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Now rerun the whole workflow. Do the logs contain what they should? Note how much easier it is to follow the progression of the workflow when the rules write to logs instead of to the terminal.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>If you have a rule with a shell directive in which several commands are run and you want to save stdout and stderr for all commands into the same log file you can add <code>exec &amp;{log}</code> as the first line of the shell directive.</p>
</div>
</div>
<p>If you run with <code>-D</code> (or <code>-S</code> for a simpler version) you will see that the summary table now also contains the log file for each of the files in the workflow.</p>
<div class="callout callout-style-default callout-note callout-titled" title="Quick recap">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Quick recap
</div>
</div>
<div class="callout-body-container callout-body">
<p>In this section we’ve learned:</p>
<ul>
<li>How to redirect output to log files with the <code>log</code> directive.</li>
</ul>
</div>
</div>
</section>
<section id="temporary-files" class="level2" data-number="7">
<h2 data-number="7" class="anchored" data-anchor-id="temporary-files"><span class="header-section-number">7</span> Temporary files</h2>
<p>It’s not uncommon that workflows contain temporary files that should be kept for some time and then deleted once they are no longer needed. A typical case could be that some operation generates a file, which is then compressed to save space or indexed to make searching faster. There is then no need to save the original output file. Take a look at the job graph for our workflow again. The output from <code>align_to_genome</code> is a BAM file, which contains information about all the reads for a sample and where they map in the genome. For downstream processing we need this file to be sorted by genome coordinates. This is what the rule <code>sort_bam</code> is for. We therefore end up with both <code>results/bam/{sample_id}.bam</code> and <code>results/bam/{sample_id}.sorted.bam</code>.</p>
<p>In Snakemake we can mark an output file as temporary like this:</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>output: temp(<span class="st">"..."</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>The file will then be deleted as soon as all jobs where it’s an input have finished. Now do this for the output of <code>align_to_genome</code>. We have to rerun the rule for it to trigger, so use <code>-R align_to_genome</code>. It should look something like this:</p>
<pre class="no-highlight"><code>.
.
rule sort_bam:
    input: results/bam/SRR935090.bam
    output: results/bam/SRR935090.sorted.bam
    jobid: 2
    wildcards: sample_id=SRR935090

Removing temporary output file results/bam/SRR935090.bam.
Finished job 2.
.
.</code></pre>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>Sometimes you may want to trigger removal of temporary files without actually rerunning the jobs. You can then use the <code>--delete-temp-output</code> flag. In some cases you may instead want to run only parts of a workflow and therefore want to prevent files marked as temporary from being deleted (because the files are needed for other parts of the workflow). In such cases you can use the <code>--notemp</code> flag.</p>
</div>
</div>
<p>Snakemake has a number of options for marking files:</p>
<ul>
<li><code>temp("...")</code>: The output file should be deleted once it’s no longer needed by any rules.</li>
<li><code>protected("...")</code>: The output file should be write-protected. Typically used to protect files that require a huge amount of computational resources from being accidentally deleted.</li>
<li><code>ancient("...")</code>: The timestamp of the input file is ignored and it’s always assumed to be older than any of the output files.</li>
<li><code>touch("...")</code>: The output file should be “touched”, <em>i.e.</em> created or updated, when the rule has finished. Typically used as “flag files” to enforce some rule execution order without real file dependencies.</li>
<li><code>directory("...")</code>: The output is a directory rather than a file.</li>
</ul>
<div class="callout callout-style-default callout-note callout-titled" title="Quick recap">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Quick recap
</div>
</div>
<div class="callout-body-container callout-body">
<p>In this section we’ve learned:</p>
<ul>
<li>How to mark an output file as temporary for automatic removal.</li>
</ul>
</div>
</div>
</section>
<section id="targets" class="level2" data-number="8">
<h2 data-number="8" class="anchored" data-anchor-id="targets"><span class="header-section-number">8</span> Targets</h2>
<p>We’ve mentioned that Snakemake rules take either strings or a list of strings as input, and that we can use any Python expression in Snakemake workflows. Here we’ll show how these features help us condense the code of rules.</p>
<p>Consider the rule <code>align_to_genome</code> below.</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>rule align_to_genome:</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""</span></span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a><span class="co">    Align a fastq file to a genome index using Bowtie 2.</span></span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>    output:</span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>        <span class="co">"results/bam/{sample_id}.bam"</span></span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a>    <span class="bu">input</span>:</span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a>        <span class="co">"data/{sample_id}.fastq.gz"</span>,</span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a>        <span class="co">"results/bowtie2/NCTC8325.1.bt2"</span>,</span>
<span id="cb32-10"><a href="#cb32-10" aria-hidden="true" tabindex="-1"></a>        <span class="co">"results/bowtie2/NCTC8325.2.bt2"</span>,</span>
<span id="cb32-11"><a href="#cb32-11" aria-hidden="true" tabindex="-1"></a>        <span class="co">"results/bowtie2/NCTC8325.3.bt2"</span>,</span>
<span id="cb32-12"><a href="#cb32-12" aria-hidden="true" tabindex="-1"></a>        <span class="co">"results/bowtie2/NCTC8325.4.bt2"</span>,</span>
<span id="cb32-13"><a href="#cb32-13" aria-hidden="true" tabindex="-1"></a>        <span class="co">"results/bowtie2/NCTC8325.rev.1.bt2"</span>,</span>
<span id="cb32-14"><a href="#cb32-14" aria-hidden="true" tabindex="-1"></a>        <span class="co">"results/bowtie2/NCTC8325.rev.2.bt2"</span></span>
<span id="cb32-15"><a href="#cb32-15" aria-hidden="true" tabindex="-1"></a>    shell:</span>
<span id="cb32-16"><a href="#cb32-16" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""</span></span>
<span id="cb32-17"><a href="#cb32-17" aria-hidden="true" tabindex="-1"></a><span class="co">        bowtie2 -x results/bowtie2/NCTC8325 -U {input[0]} &gt; {output}</span></span>
<span id="cb32-18"><a href="#cb32-18" aria-hidden="true" tabindex="-1"></a><span class="co">        """</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Here we have seven inputs; the FASTQ file with the reads and six files with similar file names from the Bowtie2 genome indexing. Instead of writing all the filenames we can tidy this up by using a Python expression to generate a list of these files instead. If you’re familiar with Python you could do this with <a href="https://docs.python.org/3/tutorial/datastructures.html#list-comprehensions">list comprehensions</a> like this:</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="bu">input</span>:</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>    <span class="co">"data/{sample_id}.fastq.gz"</span>,</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>    [<span class="ss">f"results/bowtie2/NCTC8325.</span><span class="sc">{</span>substr<span class="sc">}</span><span class="ss">.bt2"</span> <span class="cf">for</span></span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>        substr <span class="kw">in</span> [<span class="st">"1"</span>, <span class="st">"2"</span>, <span class="st">"3"</span>, <span class="st">"4"</span>, <span class="st">"rev.1"</span>, <span class="st">"rev.2"</span>]]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>This will take the elements of the list of substrings one by one, and insert that element in the place of <code>{substr}</code>. Since this type of aggregating rules are quite common, Snakemake also has a more compact way of achieving the same thing.</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="bu">input</span>:</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>    <span class="co">"data/{sample_id}.fastq.gz"</span>,</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>    expand(<span class="st">"results/bowtie2/NCTC8325.</span><span class="sc">{substr}</span><span class="st">.bt2"</span>,</span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>        substr <span class="op">=</span> [<span class="st">"1"</span>, <span class="st">"2"</span>, <span class="st">"3"</span>, <span class="st">"4"</span>, <span class="st">"rev.1"</span>, <span class="st">"rev.2"</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="callout callout-style-default callout-caution callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Caution
</div>
</div>
<div class="callout-body-container callout-body">
<p>When using expand() like this, <code>substr</code> is not a wildcard because it is resolved to the values explicitly given inside the expand expression.</p>
</div>
</div>
<p>Now change in the rules <code>index_genome</code> and <code>align_to_genome</code> to use the <code>expand()</code> expression.</p>
<p>In the workflow we decide which samples to run by including the SRR ids in the names of the inputs to the rules <code>multiqc</code> and <code>generate_count_table</code>:</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>rule generate_count_table:</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>    output:</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>        <span class="co">"results/tables/counts.tsv"</span></span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>    <span class="bu">input</span>:</span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>        bams <span class="op">=</span> [<span class="st">"results/bam/SRR935090.sorted.bam"</span>,</span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>                <span class="st">"results/bam/SRR935091.sorted.bam"</span>,</span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>                <span class="st">"results/bam/SRR935092.sorted.bam"</span>],</span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a>...</span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a>rule multiqc:</span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a>    output:</span>
<span id="cb35-11"><a href="#cb35-11" aria-hidden="true" tabindex="-1"></a>        html <span class="op">=</span> <span class="st">"results/multiqc/multiqc.html"</span>,</span>
<span id="cb35-12"><a href="#cb35-12" aria-hidden="true" tabindex="-1"></a>        stats <span class="op">=</span> <span class="st">"results/multiqc/multiqc_general_stats.txt"</span></span>
<span id="cb35-13"><a href="#cb35-13" aria-hidden="true" tabindex="-1"></a>    <span class="bu">input</span>:</span>
<span id="cb35-14"><a href="#cb35-14" aria-hidden="true" tabindex="-1"></a>        <span class="co">"results/fastqc/SRR935090_fastqc.zip"</span>,</span>
<span id="cb35-15"><a href="#cb35-15" aria-hidden="true" tabindex="-1"></a>        <span class="co">"results/fastqc/SRR935091_fastqc.zip"</span>,</span>
<span id="cb35-16"><a href="#cb35-16" aria-hidden="true" tabindex="-1"></a>        <span class="co">"results/fastqc/SRR935092_fastqc.zip"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>The output files from these two rules, <code>results/multiqc.html</code> and <code>results/tables/counts.tsv</code>, are in turn specified as input to the <code>all</code> rule at the top of the file. Because the first rule is targeted by default when we run Snakemake on the command line (like we mentioned in <a href="snakemake-4-the-mrsa-workflow">snakemake-4-the-mrsa-workflow</a>) this is what triggers the rules to run on each of the three samples.</p>
<p>However, this is a potential source of errors since it’s easy to change in one place and forget to change in the other. Because we can use Python code “everywhere” let’s instead define a list of sample ids and put at the very top of the Snakefile, just before the rule <code>all</code>:</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>SAMPLES <span class="op">=</span> [<span class="st">"SRR935090"</span>, <span class="st">"SRR935091"</span>, <span class="st">"SRR935092"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Now use <code>expand()</code> in <code>multiqc</code> and <code>generate_count_table</code> to use <code>SAMPLES</code> for the sample ids. For the <code>multiqc</code> rule it could look like this:</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="bu">input</span>:</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>    expand(<span class="st">"results/fastqc/</span><span class="sc">{sample_id}</span><span class="st">_fastqc.zip"</span>, sample_id <span class="op">=</span> SAMPLES)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>See if you can update the <code>generate_count_table</code> rule in the same manner!</p>
<div class="callout callout-style-default callout-note callout-titled" title="Quick recap">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Quick recap
</div>
</div>
<div class="callout-body-container callout-body">
<p>In this section we’ve learned:</p>
<ul>
<li>How to use the <code>expand()</code> expression to create a list with file names, inserting all provided wildcard values.</li>
</ul>
</div>
</div>
</section>
<section id="shadow-rules" class="level2" data-number="9">
<h2 data-number="9" class="anchored" data-anchor-id="shadow-rules"><span class="header-section-number">9</span> Shadow rules</h2>
<p>Take a look at the <code>index_genome</code> rule below:</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>rule index_genome:</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""</span></span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a><span class="co">    Index a genome using Bowtie 2.</span></span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a>    output:</span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a>        index <span class="op">=</span> expand(<span class="st">"results/bowtie2/NCTC8325.</span><span class="sc">{substr}</span><span class="st">.bt2"</span>,</span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a>           substr <span class="op">=</span> [<span class="st">"1"</span>, <span class="st">"2"</span>, <span class="st">"3"</span>, <span class="st">"4"</span>, <span class="st">"rev.1"</span>, <span class="st">"rev.2"</span>])</span>
<span id="cb38-8"><a href="#cb38-8" aria-hidden="true" tabindex="-1"></a>    <span class="bu">input</span>:</span>
<span id="cb38-9"><a href="#cb38-9" aria-hidden="true" tabindex="-1"></a>        <span class="co">"data/NCTC8325.fa.gz"</span></span>
<span id="cb38-10"><a href="#cb38-10" aria-hidden="true" tabindex="-1"></a>    log:</span>
<span id="cb38-11"><a href="#cb38-11" aria-hidden="true" tabindex="-1"></a>        <span class="co">"results/logs/index_genome/NCTC8325.log"</span></span>
<span id="cb38-12"><a href="#cb38-12" aria-hidden="true" tabindex="-1"></a>    shell:</span>
<span id="cb38-13"><a href="#cb38-13" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""</span></span>
<span id="cb38-14"><a href="#cb38-14" aria-hidden="true" tabindex="-1"></a><span class="co">        # Bowtie2 cannot use .gz, so unzip to a temporary file first</span></span>
<span id="cb38-15"><a href="#cb38-15" aria-hidden="true" tabindex="-1"></a><span class="co">        gunzip -c {input} &gt; tempfile</span></span>
<span id="cb38-16"><a href="#cb38-16" aria-hidden="true" tabindex="-1"></a><span class="co">        bowtie2-build tempfile results/bowtie2/NCTC8325 &gt;{log}</span></span>
<span id="cb38-17"><a href="#cb38-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-18"><a href="#cb38-18" aria-hidden="true" tabindex="-1"></a><span class="co">        # Remove the temporary file</span></span>
<span id="cb38-19"><a href="#cb38-19" aria-hidden="true" tabindex="-1"></a><span class="co">        rm tempfile</span></span>
<span id="cb38-20"><a href="#cb38-20" aria-hidden="true" tabindex="-1"></a><span class="co">        """</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>There is a temporary file here called <code>tempfile</code> which is the uncompressed version of the input, since Bowtie2 cannot use compressed files. There are a number of drawbacks with having files that aren’t explicitly part of the workflow as input/output files to rules:</p>
<ul>
<li>Snakemake cannot clean up these files if the job fails, as it would do for normal output files.</li>
<li>If several jobs are run in parallel there is a risk that they write to <code>tempfile</code> at the same time. This can lead to very scary results.</li>
<li>Sometimes we don’t know the names of all the files that a program can generate. It is, for example, not unusual that programs leave some kind of error log behind if something goes wrong.</li>
</ul>
<p>All of these issues can be dealt with by using the <code>shadow</code> option for a rule. The shadow option results in that each execution of the rule is run in an isolated temporary directory (located in <code>.snakemake/shadow/</code> by default). There are a few options for <code>shadow</code> (for the full list of these options see the <a href="https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#shadow-rules">Snakemake docs</a>). The most simple is <code>shadow: "minimal"</code>, which means that the rule is executed in an empty directory that the input files to the rule have been symlinked into. For the rule below, that means that the only file available would be <code>input.txt</code>. The shell commands would generate the files <code>some_other_junk_file</code> and <code>output.txt</code>. Lastly, Snakemake will move the output file (<code>output.txt</code>) to its “real” location and remove the whole shadow directory. We therefore never have to think about manually removing <code>some_other_junk_file</code>.</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>rule some_rule:</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>    output:</span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>        <span class="co">"output.txt"</span></span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>    <span class="bu">input</span>:</span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>        <span class="co">"input.txt"</span></span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a>    shadow: <span class="st">"minimal"</span></span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a>    shell:</span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""</span></span>
<span id="cb39-9"><a href="#cb39-9" aria-hidden="true" tabindex="-1"></a><span class="co">        touch some_other_junk_file</span></span>
<span id="cb39-10"><a href="#cb39-10" aria-hidden="true" tabindex="-1"></a><span class="co">        cp {input} {output}</span></span>
<span id="cb39-11"><a href="#cb39-11" aria-hidden="true" tabindex="-1"></a><span class="co">        """</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Try this out for the rules where we have to “manually” deal with files that aren’t tracked by Snakemake (<code>multiqc</code>, <code>index_genome</code>). Also remove the shell commands that remove temporary files from those rules, as they are no longer needed. Now rerun the workflow and validate that the temporary files don’t show up in your working directory.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>Some people use the shadow option for almost every rule and some never use it at all. One thing to keep in mind is that it leads to some extra file operations when the outputs are moved to their final location. This is no issue when the shadow directory is on the same disk as the output directory, but if you’re running on a distributed file system and generate very many or very large files it might be worth considering other options (see <em>e.g.</em> the <code>--shadow-prefix</code> flag).</p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled" title="Quick recap">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Quick recap
</div>
</div>
<div class="callout-body-container callout-body">
<p>In this section we’ve learned:</p>
<ul>
<li>How to use the shadow option to handle files that are not tracked by Snakemake.</li>
</ul>
</div>
</div>
</section>
<section id="generalising-workflows" class="level2" data-number="10">
<h2 data-number="10" class="anchored" data-anchor-id="generalising-workflows"><span class="header-section-number">10</span> Generalising workflows</h2>
<p>It’s a good idea to separate project-specific parameters from the actual implementation of the workflow. This allows anyone using the workflow to modify its behaviour without changing the underlying code, making the workflow more general.</p>
<p>In order to generalize our RNA-seq analysis workflow we should move all project-specific information to <code>config.yml</code>. This means that we want the config file to:</p>
<ul>
<li>Specify which samples to run.</li>
<li>Specify which genome to align to and where to download its sequence and annotation files.</li>
<li>(Contain any other parameters we might need to make it into a general workflow, <em>e.g.</em> to support both paired-end and single-read sequencing)</li>
</ul>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Putting all configuration in <code>config.yml</code> will break the <code>generate_rulegraph</code> rule. You can fix it either by replacing <code>--config max_reads=0</code> with <code>--configfile=config.yml</code> in the shell command of that rule in the Snakefile, or by adding <code>configfile: "config.yml"</code> to the top of the Snakefile (as mentioned in a previous tip).</p>
</div>
</div>
<p>The first point is straightforward; rather than using <code>SAMPLES = ["..."]</code> in the Snakefile we define it as a parameter in <code>config.yml</code>. You can either add it as a list similar to the way it was expressed before by adding:</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode yaml code-with-copy"><code class="sourceCode yaml"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="fu">SAMPLES</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="st">"SRR935090"</span><span class="kw">,</span><span class="at"> </span><span class="st">"SRR935091"</span><span class="kw">,</span><span class="at"> </span><span class="st">"SRR935092"</span><span class="kw">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>To <code>config.yml</code>, or you can use this YAML notation (whether you choose <code>SAMPLES</code> or <code>sample_ids</code> as the name of the entry doesn’t matter, you will just have to reference the same name in the config dictionary inside the workflow):</p>
<div class="sourceCode" id="cb41"><pre class="sourceCode yaml code-with-copy"><code class="sourceCode yaml"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sample_ids</span><span class="kw">:</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="kw">-</span><span class="at"> SRR935090</span></span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="kw">-</span><span class="at"> SRR935091</span></span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="kw">-</span><span class="at"> SRR935092</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Change the workflow to reference <code>config["sample_ids"]</code> (if using the latter example) instead of <code>SAMPLES</code>, as in:</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="fu">expand</span><span class="er">(</span><span class="st">"results/fastqc/{sample_id}_fastqc.zip"</span><span class="ex">,</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>            <span class="ex">sample_id</span> = config<span class="pp">[</span><span class="st">"sample_ids"</span><span class="pp">]</span><span class="kw">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Remove the line with <code>SAMPLES = ["SRR935090", "SRR935091", "SRR935092"]</code> that we added to the top of <code>snakefile_mrsa.smk</code> in <a href="snakemake-8-targets">Snakemake 8: Targets</a>.</p>
<p>Do a dry-run afterwards to make sure that everything works as expected.</p>
<p>You may remember from the <a href="snakemake-5-parameters">snakemake-5-parameters</a> part of this tutorial that we’re using a function to return the URL of the FASTQ files to download for each sample:</p>
<div class="sourceCode" id="cb43"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_sample_url(wildcards):</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>    samples <span class="op">=</span> {</span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>        <span class="st">"SRR935090"</span>: <span class="st">"https://figshare.scilifelab.se/ndownloader/files/39539767"</span>,</span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a>        <span class="st">"SRR935091"</span>: <span class="st">"https://figshare.scilifelab.se/ndownloader/files/39539770"</span>,</span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a>        <span class="st">"SRR935092"</span>: <span class="st">"https://figshare.scilifelab.se/ndownloader/files/39539773"</span></span>
<span id="cb43-6"><a href="#cb43-6" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb43-7"><a href="#cb43-7" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> samples[wildcards.sample_id]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Here the URLs of each sample_id is hard-coded in the <code>samples</code> dictionary inside the function. To generalize this function we can move the definition to the config file, placing it for example under an entry that we call <code>sample_urls</code> like this:</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode yaml code-with-copy"><code class="sourceCode yaml"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sample_urls</span><span class="kw">:</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">SRR935090</span><span class="kw">:</span><span class="at"> </span><span class="st">"https://figshare.scilifelab.se/ndownloader/files/39539767"</span></span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">SRR935091</span><span class="kw">:</span><span class="at"> </span><span class="st">"https://figshare.scilifelab.se/ndownloader/files/39539770"</span></span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">SRR935092</span><span class="kw">:</span><span class="at"> </span><span class="st">"https://figshare.scilifelab.se/ndownloader/files/39539773"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>This is what’s called ‘nested’ key/value pairs, meaning that each sample_id -&gt; URL pair becomes nested under the config key <code>sample_urls</code>. So in order to access the URL of <em>e.g.</em> <code>SRR935090</code> we would use <code>config["sample_urls"]["SRR935090"]</code>. This means that you will have to update the <code>get_sample_url</code> function to:</p>
<div class="sourceCode" id="cb45"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_sample_url(wildcards):</span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> config[<span class="st">"sample_urls"</span>][wildcards.sample_id]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Now the function uses the global <code>config</code> dictionary to return URLs for each sample_id. Again, do a dry-run to see that the new implementation works.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>If you were to scale up this workflow with more samples it could become impractical to have to define the URLs by hand in the config file. A tip then is to have a separate file where samples are listed in one column and the URLs (or file paths) in another column. With a few lines of python code you could then read that list at the start of the workflow and add each sample to the config dictionary.</p>
</div>
</div>
<p>Now let’s take a look at the genome reference used in the workflow. In the <code>get_genome_fasta</code> and <code>get_genome_gff3</code> rules we have hard-coded FTP paths to the FASTA GFF annotation file for the genome <code>NCTC8325</code>. We can generalize this in a similar fashion to what we did with the <code>get_SRA_by_accession</code> rule. Let’s add a nested entry called <code>genomes</code> to the config file that will hold the genome id and FTP paths to the FASTA and GFF file:</p>
<div class="sourceCode" id="cb46"><pre class="sourceCode yaml code-with-copy"><code class="sourceCode yaml"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="fu">genomes</span><span class="kw">:</span></span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">NCTC8325</span><span class="kw">:</span></span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a><span class="at">    </span><span class="fu">fasta</span><span class="kw">:</span><span class="at"> ftp://ftp.ensemblgenomes.org/pub/bacteria/release-37/fasta/bacteria_18_collection/staphylococcus_aureus_subsp_aureus_nctc_8325/dna//Staphylococcus_aureus_subsp_aureus_nctc_8325.ASM1342v1.dna_rm.toplevel.fa.gz</span></span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a><span class="at">    </span><span class="fu">gff3</span><span class="kw">:</span><span class="at"> ftp://ftp.ensemblgenomes.org/pub/bacteria/release-37/gff3/bacteria_18_collection/staphylococcus_aureus_subsp_aureus_nctc_8325//Staphylococcus_aureus_subsp_aureus_nctc_8325.ASM1342v1.37.gff3.gz</span></span>
<span id="cb46-5"><a href="#cb46-5" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">ST398</span><span class="kw">:</span></span>
<span id="cb46-6"><a href="#cb46-6" aria-hidden="true" tabindex="-1"></a><span class="at">    </span><span class="fu">fasta</span><span class="kw">:</span><span class="at"> ftp://ftp.ensemblgenomes.org/pub/bacteria/release-37/fasta/bacteria_18_collection//staphylococcus_aureus_subsp_aureus_st398/dna/Staphylococcus_aureus_subsp_aureus_st398.ASM958v1.dna.toplevel.fa.gz</span></span>
<span id="cb46-7"><a href="#cb46-7" aria-hidden="true" tabindex="-1"></a><span class="at">    </span><span class="fu">gff3</span><span class="kw">:</span><span class="at"> ftp://ftp.ensemblgenomes.org/pub/bacteria/release-37/gff3/bacteria_18_collection/staphylococcus_aureus_subsp_aureus_st398//Staphylococcus_aureus_subsp_aureus_st398.ASM958v1.37.gff3.gz</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>As you can see this is very similar to what with did with <code>sample_urls</code>, just that we have one more nested level. Now to access the FTP path to the FASTA file for genome id <code>NCTC8325</code> we can use <code>config["genomes"]["NCTC8325"]["fasta"]</code>.</p>
<p>Let’s now look at how to do the mapping from genome id to FASTA path in the rule <code>get_genome_fasta</code>. This is how the rule currently looks (if you have added the log section as previously described).</p>
<div class="sourceCode" id="cb47"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>rule get_genome_fasta:</span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""</span></span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a><span class="co">    Retrieve the sequence in fasta format for a genome.</span></span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a>    output:</span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a>        <span class="co">"data/raw_external/NCTC8325.fa.gz"</span></span>
<span id="cb47-7"><a href="#cb47-7" aria-hidden="true" tabindex="-1"></a>    log:</span>
<span id="cb47-8"><a href="#cb47-8" aria-hidden="true" tabindex="-1"></a>        <span class="co">"results/logs/get_genome_fasta/NCTC8325.log"</span></span>
<span id="cb47-9"><a href="#cb47-9" aria-hidden="true" tabindex="-1"></a>    shell:</span>
<span id="cb47-10"><a href="#cb47-10" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""</span></span>
<span id="cb47-11"><a href="#cb47-11" aria-hidden="true" tabindex="-1"></a><span class="co">        wget -o {log} ftp://ftp.ensemblgenomes.org/pub/bacteria/release-37/fasta/bacteria_18_collection/staphylococcus_aureus_subsp_aureus_nctc_8325/dna//Staphylococcus_aureus_subsp_aureus_nctc_8325.ASM1342v1.dna_rm.toplevel.fa.gz -O {output}</span></span>
<span id="cb47-12"><a href="#cb47-12" aria-hidden="true" tabindex="-1"></a><span class="co">        """</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>We don’t want the hard-coded genome id <code>NCTC8325</code>, so replace that with a wildcard, say <code>{genome_id}</code> (remember to add the wildcard to the <code>log:</code> directive as well). We now need to supply the remote paths to the FASTA file for a given genome id. Because we’ve added this information to the config file we just need to pass it to the rule in some way, and just like in the <code>get_SRA_by_accession</code> rule we’ll use a function to do the job:</p>
<div class="sourceCode" id="cb48"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_fasta_path(wildcards):</span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> config[<span class="st">"genomes"</span>][wildcards.genome_id][<span class="st">"fasta"</span>]</span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a>rule get_genome_fasta:</span>
<span id="cb48-5"><a href="#cb48-5" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""</span></span>
<span id="cb48-6"><a href="#cb48-6" aria-hidden="true" tabindex="-1"></a><span class="co">    Retrieve the sequence in fasta format for a genome.</span></span>
<span id="cb48-7"><a href="#cb48-7" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb48-8"><a href="#cb48-8" aria-hidden="true" tabindex="-1"></a>    output:</span>
<span id="cb48-9"><a href="#cb48-9" aria-hidden="true" tabindex="-1"></a>        <span class="co">"data/ref/{genome_id}.fa.gz"</span></span>
<span id="cb48-10"><a href="#cb48-10" aria-hidden="true" tabindex="-1"></a>    log:</span>
<span id="cb48-11"><a href="#cb48-11" aria-hidden="true" tabindex="-1"></a>        <span class="co">"results/logs/get_genome_fasta/{genome_id}.log"</span></span>
<span id="cb48-12"><a href="#cb48-12" aria-hidden="true" tabindex="-1"></a>    params:</span>
<span id="cb48-13"><a href="#cb48-13" aria-hidden="true" tabindex="-1"></a>        fasta_path <span class="op">=</span> get_fasta_path</span>
<span id="cb48-14"><a href="#cb48-14" aria-hidden="true" tabindex="-1"></a>    shell:</span>
<span id="cb48-15"><a href="#cb48-15" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""</span></span>
<span id="cb48-16"><a href="#cb48-16" aria-hidden="true" tabindex="-1"></a><span class="co">        wget -o {log} {params.fasta_path} -O {output}</span></span>
<span id="cb48-17"><a href="#cb48-17" aria-hidden="true" tabindex="-1"></a><span class="co">        """</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Now change the <code>get_genome_gff3</code> rule in a similar manner. Click to see the solution below if you’re having trouble.</p>
<div class="callout callout-style-default callout-tip callout-titled" title="Click to show">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-26-contents" aria-controls="callout-26" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to show
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-26" class="callout-26-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb49"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_gff_path(wildcards):</span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> config[<span class="st">"genomes"</span>][wildcards.genome_id][<span class="st">"gff3"</span>]</span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a>rule get_genome_gff3:</span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""</span></span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a><span class="co">    Retrieve annotation in gff3 format for a genome.</span></span>
<span id="cb49-7"><a href="#cb49-7" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb49-8"><a href="#cb49-8" aria-hidden="true" tabindex="-1"></a>    output:</span>
<span id="cb49-9"><a href="#cb49-9" aria-hidden="true" tabindex="-1"></a>        <span class="co">"data/ref/{genome_id}.gff3.gz"</span></span>
<span id="cb49-10"><a href="#cb49-10" aria-hidden="true" tabindex="-1"></a>    log:</span>
<span id="cb49-11"><a href="#cb49-11" aria-hidden="true" tabindex="-1"></a>        <span class="co">"results/logs/get_genome_gff3/{genome_id}.log"</span></span>
<span id="cb49-12"><a href="#cb49-12" aria-hidden="true" tabindex="-1"></a>    params:</span>
<span id="cb49-13"><a href="#cb49-13" aria-hidden="true" tabindex="-1"></a>        gff3_path <span class="op">=</span> get_gff_path</span>
<span id="cb49-14"><a href="#cb49-14" aria-hidden="true" tabindex="-1"></a>    shell:</span>
<span id="cb49-15"><a href="#cb49-15" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""</span></span>
<span id="cb49-16"><a href="#cb49-16" aria-hidden="true" tabindex="-1"></a><span class="co">        wget -o {log} {params.gff3_path} -O {output}</span></span>
<span id="cb49-17"><a href="#cb49-17" aria-hidden="true" tabindex="-1"></a><span class="co">        """</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<p>Also change in <code>index_genome</code> to use a wildcard rather than a hard-coded genome id. Here you will run into a complication if you have followed the previous instructions and use the <code>expand()</code> expression. We want the list to expand to <code>["results/bowtie2/{genome_id}.1.bt2", "results/bowtie2/{genome_id}.2.bt2", ...]</code>, <em>i.e.</em> only expanding the wildcard referring to the Bowtie2 index. To keep the <code>genome_id</code> wildcard from being expanded we have to “mask” it with double curly brackets: <code>{genome_id}</code>. In addition, we need to replace the hard-coded <code>results/bowtie2/NCTC8325</code> in the shell directive of the rule with the genome id wildcard. Inside the shell directive the wildcard object is accessed with this syntax: <code>{wildcards.genome_id}</code>, so the Bowtie2-build command should be:</p>
<div class="sourceCode" id="cb50"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bowtie2-build</span> tempfile results/bowtie2/{wildcards.genome_id} <span class="op">&gt;</span> {log}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Note that this will only work if the <code>{genome_id}</code> wildcard can be resolved to something defined in the config (currently <code>NCTC8325</code> or <code>ST398</code>). If you try to generate a FASTA file for a genome id not defined in the config Snakemake will complain, even at the dry-run stage.</p>
<p>Finally, remember that any wildcards need to be present both in the <code>output:</code> and <code>log:</code> directives? This means we have to update the <code>log:</code> directive in <code>index_genome</code> as well. The final rule should look like this:</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>rule index_genome:</span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""</span></span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a><span class="co">    Index a genome using Bowtie 2.</span></span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb51-5"><a href="#cb51-5" aria-hidden="true" tabindex="-1"></a>    output:</span>
<span id="cb51-6"><a href="#cb51-6" aria-hidden="true" tabindex="-1"></a>        expand(<span class="st">"results/bowtie2/</span><span class="sc">{{</span><span class="st">genome_id</span><span class="sc">}}</span><span class="st">.</span><span class="sc">{substr}</span><span class="st">.bt2"</span>,</span>
<span id="cb51-7"><a href="#cb51-7" aria-hidden="true" tabindex="-1"></a>            substr <span class="op">=</span> [<span class="st">"1"</span>, <span class="st">"2"</span>, <span class="st">"3"</span>, <span class="st">"4"</span>, <span class="st">"rev.1"</span>, <span class="st">"rev.2"</span>])</span>
<span id="cb51-8"><a href="#cb51-8" aria-hidden="true" tabindex="-1"></a>    <span class="bu">input</span>:</span>
<span id="cb51-9"><a href="#cb51-9" aria-hidden="true" tabindex="-1"></a>        <span class="co">"data/ref/{genome_id}.fa.gz"</span></span>
<span id="cb51-10"><a href="#cb51-10" aria-hidden="true" tabindex="-1"></a>    log:</span>
<span id="cb51-11"><a href="#cb51-11" aria-hidden="true" tabindex="-1"></a>        <span class="co">"results/logs/index_genome/{genome_id}.log"</span></span>
<span id="cb51-12"><a href="#cb51-12" aria-hidden="true" tabindex="-1"></a>    shadow: <span class="st">"minimal"</span></span>
<span id="cb51-13"><a href="#cb51-13" aria-hidden="true" tabindex="-1"></a>    shell:</span>
<span id="cb51-14"><a href="#cb51-14" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""</span></span>
<span id="cb51-15"><a href="#cb51-15" aria-hidden="true" tabindex="-1"></a><span class="co">        # Bowtie2 cannot use .gz, so unzip to a temporary file first</span></span>
<span id="cb51-16"><a href="#cb51-16" aria-hidden="true" tabindex="-1"></a><span class="co">        gunzip -c {input} &gt; tempfile</span></span>
<span id="cb51-17"><a href="#cb51-17" aria-hidden="true" tabindex="-1"></a><span class="co">        bowtie2-build tempfile results/bowtie2/{wildcards.genome_id} &gt; {log}</span></span>
<span id="cb51-18"><a href="#cb51-18" aria-hidden="true" tabindex="-1"></a><span class="co">        """</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Good job! The rules <code>get_genome_fasta</code>, <code>get_genome_gff3</code> and <code>index_genome</code> can now download and index <em>any genome</em> as long as we provide valid links in the config file.</p>
<p>However, we need to define somewhere which genome id we actually want to use when running the workflow. This needs to be done both in <code>align_to_genome</code> and <code>generate_count_table</code>. Do this by introducing a parameter in <code>config.yml</code> called <code>"genome_id"</code> (you can set it to either <code>NCTC8325</code> or <code>ST398</code>), <em>e.g.</em>:</p>
<div class="sourceCode" id="cb52"><pre class="sourceCode yaml code-with-copy"><code class="sourceCode yaml"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="fu">genome_id</span><span class="kw">:</span><span class="at"> </span><span class="st">"NCTC8325"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Now we can resolve the <code>genome_id</code> wildcard from the config. See below for an example for <code>align_to_genome</code>. Here the <code>substr</code> wildcard gets expanded from a list while <code>genome_id</code> gets expanded from the config file.</p>
<div class="sourceCode" id="cb53"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="bu">input</span>:</span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>    <span class="co">"data/{sample_id}.fastq.gz"</span>,</span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a>    index <span class="op">=</span> expand(<span class="st">"results/bowtie2/</span><span class="sc">{genome_id}</span><span class="st">.</span><span class="sc">{substr}</span><span class="st">.bt2"</span>,</span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a>           genome_id <span class="op">=</span> config[<span class="st">"genome_id"</span>],</span>
<span id="cb53-5"><a href="#cb53-5" aria-hidden="true" tabindex="-1"></a>           substr <span class="op">=</span> [<span class="st">"1"</span>, <span class="st">"2"</span>, <span class="st">"3"</span>, <span class="st">"4"</span>, <span class="st">"rev.1"</span>, <span class="st">"rev.2"</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Also change the hard-coded genome id in the <code>generate_count_table</code> input in a similar manner:</p>
<div class="sourceCode" id="cb54"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>rule generate_count_table:</span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""</span></span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a><span class="co">    Generate a count table using featureCounts.</span></span>
<span id="cb54-4"><a href="#cb54-4" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb54-5"><a href="#cb54-5" aria-hidden="true" tabindex="-1"></a>    output:</span>
<span id="cb54-6"><a href="#cb54-6" aria-hidden="true" tabindex="-1"></a>        <span class="co">"results/tables/counts.tsv"</span>,</span>
<span id="cb54-7"><a href="#cb54-7" aria-hidden="true" tabindex="-1"></a>        <span class="co">"results/tables/counts.tsv.summary"</span></span>
<span id="cb54-8"><a href="#cb54-8" aria-hidden="true" tabindex="-1"></a>    <span class="bu">input</span>:</span>
<span id="cb54-9"><a href="#cb54-9" aria-hidden="true" tabindex="-1"></a>        bams<span class="op">=</span>expand(<span class="st">"results/bam/</span><span class="sc">{sample_id}</span><span class="st">.sorted.bam"</span>,</span>
<span id="cb54-10"><a href="#cb54-10" aria-hidden="true" tabindex="-1"></a>                    sample_id <span class="op">=</span> config[<span class="st">"sample_ids"</span>]),</span>
<span id="cb54-11"><a href="#cb54-11" aria-hidden="true" tabindex="-1"></a>        annotation<span class="op">=</span>expand(<span class="st">"data/ref/</span><span class="sc">{genome_id}</span><span class="st">.gff3.gz"</span>,</span>
<span id="cb54-12"><a href="#cb54-12" aria-hidden="true" tabindex="-1"></a>                    genome_id <span class="op">=</span> config[<span class="st">"genome_id"</span>])</span>
<span id="cb54-13"><a href="#cb54-13" aria-hidden="true" tabindex="-1"></a>    log:</span>
<span id="cb54-14"><a href="#cb54-14" aria-hidden="true" tabindex="-1"></a>        <span class="co">"results/logs/generate_count_table.log"</span></span>
<span id="cb54-15"><a href="#cb54-15" aria-hidden="true" tabindex="-1"></a>    shell:</span>
<span id="cb54-16"><a href="#cb54-16" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""</span></span>
<span id="cb54-17"><a href="#cb54-17" aria-hidden="true" tabindex="-1"></a><span class="co">        featureCounts -t gene -g gene_id -a {input.annotation} -o {output[0]} {input.bams} 2&gt;{log}</span></span>
<span id="cb54-18"><a href="#cb54-18" aria-hidden="true" tabindex="-1"></a><span class="co">        """</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>In general, we want the rules as far downstream as possible in the workflow to be the ones that determine what the wildcards should resolve to. In our case this is <code>align_to_genome</code> and <code>generate_count_table</code>. You can think of it like the rule that really “needs” the file asks for it, and then it’s up to Snakemake to determine how it can use all the available rules to generate it. Here the <code>align_to_genome</code> rule says “I need this genome index to align my sample to” and then it’s up to Snakemake to determine how to download and build the index.</p>
<p>One last thing is to change the hard-coded <code>NCTC8325</code> in the <code>shell:</code> directive of <code>align_to_genome</code>. Bowtie2 expects the index name supplied with the <code>-x</code> flag to be without the “.*.bt2” suffix so we can’t use <code>-x {input.index}</code>. Instead we’ll insert the genome_id directly from the config like this:</p>
<div class="sourceCode" id="cb55"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="ex">shell:</span></span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>    <span class="st">"""</span></span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a><span class="st">    bowtie2 -x results/bowtie2/{config[genome_id]} -U {input[0]} &gt; {output} 2&gt;{log}</span></span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a><span class="st">    """</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="callout callout-style-default callout-note callout-titled" title="Summary">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Summary
</div>
</div>
<div class="callout-body-container callout-body">
<p>Well done! You now have a complete Snakemake workflow with a number of excellent features:</p>
<ul>
<li>A general RNA-seq pipeline which can easily be reused between projects, thanks to clear separation between code and settings.</li>
<li>Great traceability due to logs and summary tables.</li>
<li>Clearly defined the environment for the workflow using Conda.</li>
<li>The workflow is neat and free from temporary files due to using <code>temp()</code> and <code>shadow</code>.</li>
<li>A logical directory structure which makes it easy to separate data and results of different software packages.</li>
<li>A project set up in a way that makes it very easy to distribute and reproduce either via Git, Snakemake’s <code>--archive</code> option or a Docker image.</li>
</ul>
</div>
</div>
</section>
<section id="reading-samples-from-a-file-instead-of-hard-coding-them" class="level2" data-number="11">
<h2 data-number="11" class="anchored" data-anchor-id="reading-samples-from-a-file-instead-of-hard-coding-them"><span class="header-section-number">11</span> Reading samples from a file instead of hard-coding them</h2>
<p>So far we’ve specified the samples to use in the workflow either as a hard-coded list in the Snakefile, or as a list in the configuration file. This is of course impractical for large real-world examples. Here we’ll just quickly show how you could supply the samples instead via a tab-separated file. For example you could create a file called <code>samples.tsv</code> with the following content:</p>
<pre><code>SRR935090   https://figshare.scilifelab.se/ndownloader/files/39539767
SRR935091   https://figshare.scilifelab.se/ndownloader/files/39539770
SRR935092   https://figshare.scilifelab.se/ndownloader/files/39539773</code></pre>
<p>The first column has the sample id and the second column has the url to the fastq file. Now in order to read this into the workflow we need to use a few lines of python code. Since you can mix python code with rule definitions in Snakemake we’ll just add the following lines to the top of the Snakefile:</p>
<div class="sourceCode" id="cb57"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="co"># define an empty 'samples' dictionary</span></span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>samples <span class="op">=</span> {}</span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a><span class="co"># read the sample list file and populate the dictionary</span></span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> <span class="bu">open</span>(<span class="st">"samples.tsv"</span>, <span class="st">"r"</span>) <span class="im">as</span> fhin:</span>
<span id="cb57-5"><a href="#cb57-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> line <span class="kw">in</span> fhin:</span>
<span id="cb57-6"><a href="#cb57-6" aria-hidden="true" tabindex="-1"></a>        <span class="co"># strip the newline character from the end of the line</span></span>
<span id="cb57-7"><a href="#cb57-7" aria-hidden="true" tabindex="-1"></a>        <span class="co"># then split by tab character to get the sample id and url</span></span>
<span id="cb57-8"><a href="#cb57-8" aria-hidden="true" tabindex="-1"></a>        sample_id, url <span class="op">=</span> line.strip().split(<span class="st">"</span><span class="ch">\t</span><span class="st">"</span>)</span>
<span id="cb57-9"><a href="#cb57-9" aria-hidden="true" tabindex="-1"></a>        <span class="co"># store the url in the dictionary with the sample id as key</span></span>
<span id="cb57-10"><a href="#cb57-10" aria-hidden="true" tabindex="-1"></a>        samples[sample_id] <span class="op">=</span> url</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Now we can use the <code>samples</code> dictionary in the workflow. For example, to get the url for <code>SRR935090</code> we can use <code>samples["SRR935090"]</code>.</p>
<p>For example, the <code>get_sample_url</code> function can now be written as:</p>
<div class="sourceCode" id="cb58"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_sample_url(wildcards):</span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> samples[wildcards.sample_id]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>We can also use the <code>samples</code> dictionary in <code>expand()</code>, for example in the <code>multiqc</code> rule:</p>
<div class="sourceCode" id="cb59"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>rule multiqc:</span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""</span></span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a><span class="co">    Aggregate all FastQC reports into a MultiQC report.</span></span>
<span id="cb59-4"><a href="#cb59-4" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb59-5"><a href="#cb59-5" aria-hidden="true" tabindex="-1"></a>    output:</span>
<span id="cb59-6"><a href="#cb59-6" aria-hidden="true" tabindex="-1"></a>        html<span class="op">=</span><span class="st">"results/multiqc/multiqc.html"</span>,</span>
<span id="cb59-7"><a href="#cb59-7" aria-hidden="true" tabindex="-1"></a>        stats<span class="op">=</span><span class="st">"results/multiqc/multiqc_general_stats.txt"</span></span>
<span id="cb59-8"><a href="#cb59-8" aria-hidden="true" tabindex="-1"></a>    <span class="bu">input</span>:</span>
<span id="cb59-9"><a href="#cb59-9" aria-hidden="true" tabindex="-1"></a>        expand(<span class="st">"results/fastqc/</span><span class="sc">{sample_id}</span><span class="st">_fastqc.zip"</span>, sample_id <span class="op">=</span> samples.keys())</span>
<span id="cb59-10"><a href="#cb59-10" aria-hidden="true" tabindex="-1"></a>    log:</span>
<span id="cb59-11"><a href="#cb59-11" aria-hidden="true" tabindex="-1"></a>        <span class="co">"results/logs/multiqc/multiqc.log"</span></span>
<span id="cb59-12"><a href="#cb59-12" aria-hidden="true" tabindex="-1"></a>    shadow: <span class="st">"minimal"</span></span>
<span id="cb59-13"><a href="#cb59-13" aria-hidden="true" tabindex="-1"></a>    shell:</span>
<span id="cb59-14"><a href="#cb59-14" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""</span></span>
<span id="cb59-15"><a href="#cb59-15" aria-hidden="true" tabindex="-1"></a><span class="co">        # Run multiQC and keep the html report</span></span>
<span id="cb59-16"><a href="#cb59-16" aria-hidden="true" tabindex="-1"></a><span class="co">        multiqc -n multiqc.html {input} 2&gt; {log}</span></span>
<span id="cb59-17"><a href="#cb59-17" aria-hidden="true" tabindex="-1"></a><span class="co">        mv multiqc.html {output.html}</span></span>
<span id="cb59-18"><a href="#cb59-18" aria-hidden="true" tabindex="-1"></a><span class="co">        mv multiqc_data/multiqc_general_stats.txt {output.stats}</span></span>
<span id="cb59-19"><a href="#cb59-19" aria-hidden="true" tabindex="-1"></a><span class="co">        """</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Now this depends on there being a <code>samples.tsv</code> file in the working directory. To make this a configurable parameter we can add it to the config file:</p>
<div class="sourceCode" id="cb60"><pre class="sourceCode yaml code-with-copy"><code class="sourceCode yaml"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sample_list</span><span class="kw">:</span><span class="at"> </span><span class="st">"samples.tsv"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>and update the code for populating the <code>samples</code> dictionary:</p>
<div class="sourceCode" id="cb61"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="co"># define an empty 'samples' dictionary</span></span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a>samples <span class="op">=</span> {}</span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a><span class="co"># read the sample list file and populate the dictionary</span></span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> <span class="bu">open</span>(config[<span class="st">"sample_list"</span>], <span class="st">"r"</span>) <span class="im">as</span> fhin:</span>
<span id="cb61-5"><a href="#cb61-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> line <span class="kw">in</span> fhin:</span>
<span id="cb61-6"><a href="#cb61-6" aria-hidden="true" tabindex="-1"></a>        <span class="co"># strip the newline character from the end of the line</span></span>
<span id="cb61-7"><a href="#cb61-7" aria-hidden="true" tabindex="-1"></a>        <span class="co"># then split by tab character to get the sample id and url</span></span>
<span id="cb61-8"><a href="#cb61-8" aria-hidden="true" tabindex="-1"></a>        sample_id, url <span class="op">=</span> line.strip().split(<span class="st">"</span><span class="ch">\t</span><span class="st">"</span>)</span>
<span id="cb61-9"><a href="#cb61-9" aria-hidden="true" tabindex="-1"></a>        <span class="co"># store the url in the dictionary with the sample id as key</span></span>
<span id="cb61-10"><a href="#cb61-10" aria-hidden="true" tabindex="-1"></a>        samples[sample_id] <span class="op">=</span> url</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>This way, anyone can take our Snakefile and just update the path to their own <code>sample_list</code> using the config file.</p>
<div class="callout callout-style-default callout-note callout-titled" title="Quick recap">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Quick recap
</div>
</div>
<div class="callout-body-container callout-body">
<p>In this section we’ve learned:</p>
<ul>
<li>How to generalize a Snakemake workflow.</li>
</ul>
</div>
</div>
</section>
<section id="extra-material" class="level2" data-number="12">
<h2 data-number="12" class="anchored" data-anchor-id="extra-material"><span class="header-section-number">12</span> Extra material</h2>
<p>If you want to read more about Snakemake in general you can find several resources here:</p>
<ul>
<li>The Snakemake documentation is available on <a href="https://snakemake.readthedocs.io/en/stable/#">ReadTheDocs</a>.</li>
<li>Here is another (quite in-depth) <a href="https://snakemake.readthedocs.io/en/stable/tutorial/tutorial.html#tutorial">tutorial</a>.</li>
<li>If you have questions, check out <a href="https://stackoverflow.com/questions/tagged/snakemake">stack overflow</a>.</li>
</ul>
<section id="using-containers-in-snakemake" class="level3" data-number="12.1">
<h3 data-number="12.1" class="anchored" data-anchor-id="using-containers-in-snakemake"><span class="header-section-number">12.1</span> Using containers in Snakemake</h3>
<p>Snakemake also supports defining an Apptainer or Docker container for each rule (you will have time to work on the <a href="containers-1-introduction">Containers tutorial</a> later during the course). Analogous to using a rule-specific Conda environment, specify <code>container: "docker://some-account/rule-specific-image"</code> in the rule definition. Instead of a link to a container image, it is also possible to provide the path to a <code>*.sif</code> file (= a <em>Singularity image file</em>). When executing Snakemake, add the <code>--software-deployment-method apptainer</code> (or the shorthand <code>--sdm apptainer</code>) flag to the command line. For the given rule, an Apptainer container will then be created from the image or file that is provided in the rule definition on the fly by Snakemake and the rule will be run in this container.</p>
<p>You can find pre-made Apptainer or Docker images for many tools on <a href="https://biocontainers.pro/">https://biocontainers.pro/</a> (bioinformatics-specific) or on <a href="https://hub.docker.com/">https://hub.docker.com/</a>.</p>
<p>Here is an example for a rule and its execution:</p>
<div class="sourceCode" id="cb62"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a>rule align_to_genome:</span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a>    output:</span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a>        temp(<span class="st">"results/bam/{sample_id,\w+}.bam"</span>)</span>
<span id="cb62-4"><a href="#cb62-4" aria-hidden="true" tabindex="-1"></a>    <span class="bu">input</span>:</span>
<span id="cb62-5"><a href="#cb62-5" aria-hidden="true" tabindex="-1"></a>        fastq <span class="op">=</span> <span class="st">"data/</span><span class="sc">{sample_id}</span><span class="st">.fastq.gz"</span>,</span>
<span id="cb62-6"><a href="#cb62-6" aria-hidden="true" tabindex="-1"></a>        index <span class="op">=</span> expand(<span class="st">"results/bowtie2/</span><span class="sc">{genome_id}</span><span class="st">.</span><span class="sc">{substr}</span><span class="st">.bt2"</span>,</span>
<span id="cb62-7"><a href="#cb62-7" aria-hidden="true" tabindex="-1"></a>            genome_id<span class="op">=</span>config[<span class="st">"genome_id"</span>],</span>
<span id="cb62-8"><a href="#cb62-8" aria-hidden="true" tabindex="-1"></a>            substr<span class="op">=</span>[<span class="st">"1"</span>, <span class="st">"2"</span>, <span class="st">"3"</span>, <span class="st">"4"</span>, <span class="st">"rev.1"</span>, <span class="st">"rev.2"</span>])</span>
<span id="cb62-9"><a href="#cb62-9" aria-hidden="true" tabindex="-1"></a>    log:</span>
<span id="cb62-10"><a href="#cb62-10" aria-hidden="true" tabindex="-1"></a>        expand(<span class="st">"results/logs/align_to_genome/</span><span class="sc">{{</span><span class="st">sample_id</span><span class="sc">}}</span><span class="st">_</span><span class="sc">{genome_id}</span><span class="st">.log"</span>,</span>
<span id="cb62-11"><a href="#cb62-11" aria-hidden="true" tabindex="-1"></a>            genome_id <span class="op">=</span> config[<span class="st">"genome_id"</span>])</span>
<span id="cb62-12"><a href="#cb62-12" aria-hidden="true" tabindex="-1"></a>    container: <span class="st">"docker://quay.io/biocontainers/bowtie2:2.5.0--py310h8d7afc0_0"</span></span>
<span id="cb62-13"><a href="#cb62-13" aria-hidden="true" tabindex="-1"></a>    shell:</span>
<span id="cb62-14"><a href="#cb62-14" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""</span></span>
<span id="cb62-15"><a href="#cb62-15" aria-hidden="true" tabindex="-1"></a><span class="co">        bowtie2 -x results/bowtie2/{config[genome_id]} -U {input.fastq} &gt; {output} 2&gt;{log}</span></span>
<span id="cb62-16"><a href="#cb62-16" aria-hidden="true" tabindex="-1"></a><span class="co">        """</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Start your Snakemake workflow with the following command:</p>
<div class="sourceCode" id="cb63"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="ex">snakemake</span> <span class="at">--software-deployment-method</span> apptainer</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Feel free to modify the MRSA workflow according to this example. As Apptainer is a container software that was developed for HPC clusters, and for example the Mac version is still a beta version, it might not work to run your updated Snakemake workflow with Apptainer locally on your computer. In the next section we explain how you can run Snakemake workflows on UPPMAX where Apptainer is pre-installed.</p>
</section>
<section id="running-snakemake-workflows-on-hpc-clusters" class="level3" data-number="12.2">
<h3 data-number="12.2" class="anchored" data-anchor-id="running-snakemake-workflows-on-hpc-clusters"><span class="header-section-number">12.2</span> Running Snakemake workflows on HPC clusters</h3>
<p>If you need to run a Snakemake workflow on a high-performance computing (HPC) cluster you have a wide range of options at your disposal. Via the <a href="https://snakemake.github.io/snakemake-plugin-catalog/">plugin catalog</a> you can find plugins that will add support for various HPC schedulers to Snakemake.</p>
<p>Here we will focus on how to run Snakemake workflows on clusters with SLURM, a workload manager commonly used on HPC clusters in Sweden such as <a href="https://www.uu.se/centrum/uppmax/resurser/kluster/rackham">Rackham</a>, <a href="https://www.nsc.liu.se/systems/tetralith/">Tetralith</a> and <a href="https://www.pdc.kth.se/hpc-services/computing-systems">Dardel</a>.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>When running on remote clusters we highly recommend to use a session manager like <a href="https://github.com/tmux/tmux/wiki">tmux</a> or <a href="https://www.gnu.org/software/screen/manual/screen.html#Overview">screen</a> so that you can run your workflow in a session in the background while doing other things on the cluster or even logging out of the cluster.</p>
</div>
</div>
<section id="option-1-run-the-entire-workflow-as-a-single-job" class="level4" data-number="12.2.1">
<h4 data-number="12.2.1" class="anchored" data-anchor-id="option-1-run-the-entire-workflow-as-a-single-job"><span class="header-section-number">12.2.1</span> Option 1: Run the entire workflow as a single job</h4>
<p>For short workflows with only a few rules that need the same compute resources in terms of CPU (cores) and memory, you can submit the entire workflow as a job directly to the SLURM scheduler, or start an interactive job (in your <code>tmux</code> or <code>screen</code> session) and run your Snakemake workflow as you would do that on your local machine. Make sure to give your job enough time to finish running all rules of your Snakemake workflow.</p>
<p>If you choose this option, you don’t need to install anything from the plugin catalogue. However, your workflow may not run as efficiently as it could if you were to add SLURM support in Snakemake.</p>
</section>
<section id="option-2-use-built-in-slurm-support" class="level4" data-number="12.2.2">
<h4 data-number="12.2.2" class="anchored" data-anchor-id="option-2-use-built-in-slurm-support"><span class="header-section-number">12.2.2</span> Option 2: Use built-in SLURM support</h4>
<p>For workflows with long run times and/or where each rule requires different compute resources, Snakemake comes with built in functionality for interacting with the SLURM workload manager and send each rule as a job to the SLURM queue and to track the status of each job.</p>
<p>In this case, you can start the workflow on the login node and let it run there until all jobs have finished. Given that workflows often consist of many rules, some of which may be highly resource demanding, this is the option we recommend when running most Snakemake workflows on HPC clusters.</p>
<p>To add SLURM support to Snakemake you first need to install the <a href="https://snakemake.github.io/snakemake-plugin-catalog/plugins/executor/slurm.html">SLURM plugin</a> from the plugin catalog. This can be done with conda:</p>
<div class="sourceCode" id="cb64"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> install <span class="at">-c</span> conda-forge snakemake-executor-plugin-slurm</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Once installed, adding the <code>--executor slurm</code> flag to your Snakemake command line call will enable the plugin. You also need to specify how many jobs Snakemake can submit to the SLURM queue at the same time with the <code>-j</code> flag. For example, to allow up to 100 jobs to be put into the queue at any given time, you would run Snakemake with the following command:</p>
<div class="sourceCode" id="cb65"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="ex">snakemake</span> <span class="at">--executor</span> slurm <span class="at">-j</span> 100 <span class="op">&lt;</span>other flags<span class="op">&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
</section>
<section id="specifying-resources-for-slurm" class="level3" data-number="12.3">
<h3 data-number="12.3" class="anchored" data-anchor-id="specifying-resources-for-slurm"><span class="header-section-number">12.3</span> Specifying resources for SLURM</h3>
<p>Depending on the cluster you are using, you will need to specify some resource requirements for the rules in your workflow, such as the number of CPUs, memory, runtime and account id. This can be done either:</p>
<ol type="1">
<li>directly on the command line with the <code>--default-resources</code> flag which sets default resource settings for all rules</li>
<li>in the rule definition of your workflow using the <code>resources:</code> directive, or</li>
<li>in a <a href="https://snakemake.readthedocs.io/en/stable/executing/cli.html#profiles">configuration profile</a>, a folder with a <code>config.yaml</code> file that contains the resource settings.</li>
</ol>
<p>You can also use a combination of these methods. For example, the SLURM account id (_e.g.&nbsp;<code>naiss-2023-01-001</code>), which will most likely be the same for all rules, can be set with <code>--default-resources</code>:</p>
<div class="sourceCode" id="cb66"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="ex">snakemake</span> <span class="at">--executor</span> slurm <span class="at">-j</span> 100 <span class="at">--default-resources</span> slurm_account=naiss-2023-01-001</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Rule-specific resources such as runtime, memory and number of CPUs can be set in the rule definition, for example:</p>
<div class="sourceCode" id="cb67"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a>rule testrule:</span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a>    output:</span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a>        <span class="co">"results/output.txt"</span></span>
<span id="cb67-4"><a href="#cb67-4" aria-hidden="true" tabindex="-1"></a>    resources:</span>
<span id="cb67-5"><a href="#cb67-5" aria-hidden="true" tabindex="-1"></a>        runtime <span class="op">=</span> <span class="dv">60</span>,</span>
<span id="cb67-6"><a href="#cb67-6" aria-hidden="true" tabindex="-1"></a>        mem_mb <span class="op">=</span> <span class="dv">16000</span>,</span>
<span id="cb67-7"><a href="#cb67-7" aria-hidden="true" tabindex="-1"></a>        cpus_per_task <span class="op">=</span> <span class="dv">4</span></span>
<span id="cb67-8"><a href="#cb67-8" aria-hidden="true" tabindex="-1"></a>    shell:</span>
<span id="cb67-9"><a href="#cb67-9" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""</span></span>
<span id="cb67-10"><a href="#cb67-10" aria-hidden="true" tabindex="-1"></a><span class="co">        uname -a &gt; {output}</span></span>
<span id="cb67-11"><a href="#cb67-11" aria-hidden="true" tabindex="-1"></a><span class="co">        """</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>This rule uses the standard resource <code>runtime</code> to set the maximum allowed time (in minutes) for the rule, sets the memory requirement with <code>mem_mb</code> and the number of requested CPUs with <code>cpus_per_task</code>. In this example the rule will have a time limit of 60 minutes, will require 16G of RAM and 4 CPUs.</p>
<p>Some clusters also require you to specify the <strong>partition</strong> you want to run your job on. The partition name will differ between clusters, for example the Rackham cluster uses <code>core</code> and <code>node</code> partitions, while Dardel uses <em>e.g.</em> <code>shared</code> and <code>main</code>. See the documentation for the cluster you are using for more information.</p>
<p>The partition can be set with the <code>slurm_partition</code> resource, for example like so:</p>
<div class="sourceCode" id="cb68"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a>rule testrule:</span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a>    output:</span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a>        <span class="co">"results/output.txt"</span></span>
<span id="cb68-4"><a href="#cb68-4" aria-hidden="true" tabindex="-1"></a>    resources:</span>
<span id="cb68-5"><a href="#cb68-5" aria-hidden="true" tabindex="-1"></a>        runtime <span class="op">=</span> <span class="dv">60</span>,</span>
<span id="cb68-6"><a href="#cb68-6" aria-hidden="true" tabindex="-1"></a>        mem_mb <span class="op">=</span> <span class="dv">16000</span>,</span>
<span id="cb68-7"><a href="#cb68-7" aria-hidden="true" tabindex="-1"></a>        cpus_per_task <span class="op">=</span> <span class="dv">4</span>,</span>
<span id="cb68-8"><a href="#cb68-8" aria-hidden="true" tabindex="-1"></a>        slurm_partition: <span class="st">"shared"</span></span>
<span id="cb68-9"><a href="#cb68-9" aria-hidden="true" tabindex="-1"></a>    shell:</span>
<span id="cb68-10"><a href="#cb68-10" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""</span></span>
<span id="cb68-11"><a href="#cb68-11" aria-hidden="true" tabindex="-1"></a><span class="co">        uname -a &gt; {output}</span></span>
<span id="cb68-12"><a href="#cb68-12" aria-hidden="true" tabindex="-1"></a><span class="co">        """</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>To make it easy to adapt your workflow to different compute clusters it is recommended to define resource settings in a <strong>configuration profile</strong>. A configuration profile is a folder with a <code>config.yaml</code> file that contains values for Snakemake command line arguments, allowing you to modify the behavior of Snakemake without changing the workflow code. For example, you could create a <code>dardel</code> folder (<em>e.g.</em> in the root of your workflow) with a <code>config.yaml</code> file that contains the following:</p>
<div class="sourceCode" id="cb69"><pre class="sourceCode yaml code-with-copy"><code class="sourceCode yaml"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="fu">executor</span><span class="kw">:</span><span class="at"> </span><span class="st">"slurm"</span></span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a><span class="fu">jobs</span><span class="kw">:</span><span class="at"> </span><span class="dv">100</span></span>
<span id="cb69-3"><a href="#cb69-3" aria-hidden="true" tabindex="-1"></a><span class="fu">default-resources</span><span class="kw">:</span></span>
<span id="cb69-4"><a href="#cb69-4" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">slurm_account</span><span class="kw">:</span><span class="at"> </span><span class="st">"naiss-2023-01-001"</span></span>
<span id="cb69-5"><a href="#cb69-5" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">slurm_partition</span><span class="kw">:</span><span class="at"> </span><span class="st">"shared"</span></span>
<span id="cb69-6"><a href="#cb69-6" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">mem_mb</span><span class="kw">:</span><span class="at"> </span><span class="dv">16000</span></span>
<span id="cb69-7"><a href="#cb69-7" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">cpus_per_task</span><span class="kw">:</span><span class="at"> </span><span class="dv">4</span></span>
<span id="cb69-8"><a href="#cb69-8" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">runtime</span><span class="kw">:</span><span class="at"> </span><span class="dv">60</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>This yaml-formatted file contains Snakemake command line arguments that will be used when running the workflow. You can then run Snakemake with the <code>--profile</code> flag pointing to the folder containing the <code>config.yaml</code> file:</p>
<div class="sourceCode" id="cb70"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a><span class="ex">snakemake</span> <span class="at">--profile</span> dardel</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>This greatly simplifies running the workflow on different clusters, and makes the command line call much more succinct.</p>
<p>To set rule-specific resources in the configuration profile, you can add a <code>set_resources:</code> section to the <code>config.yaml</code> file:</p>
<div class="sourceCode" id="cb71"><pre class="sourceCode yaml code-with-copy"><code class="sourceCode yaml"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="fu">executor</span><span class="kw">:</span><span class="at"> </span><span class="st">"slurm"</span></span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a><span class="fu">jobs</span><span class="kw">:</span><span class="at"> </span><span class="dv">100</span></span>
<span id="cb71-3"><a href="#cb71-3" aria-hidden="true" tabindex="-1"></a><span class="fu">default-resources</span><span class="kw">:</span></span>
<span id="cb71-4"><a href="#cb71-4" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">slurm_account</span><span class="kw">:</span><span class="at"> </span><span class="st">"naiss-2023-01-001"</span></span>
<span id="cb71-5"><a href="#cb71-5" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">slurm_partition</span><span class="kw">:</span><span class="at"> </span><span class="st">"shared"</span></span>
<span id="cb71-6"><a href="#cb71-6" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">mem_mb</span><span class="kw">:</span><span class="at"> </span><span class="dv">16000</span></span>
<span id="cb71-7"><a href="#cb71-7" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">cpus_per_task</span><span class="kw">:</span><span class="at"> </span><span class="dv">4</span></span>
<span id="cb71-8"><a href="#cb71-8" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">runtime</span><span class="kw">:</span><span class="at"> </span><span class="dv">60</span></span>
<span id="cb71-9"><a href="#cb71-9" aria-hidden="true" tabindex="-1"></a><span class="fu">set_resources</span><span class="kw">:</span></span>
<span id="cb71-10"><a href="#cb71-10" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">index_genome</span><span class="kw">:</span></span>
<span id="cb71-11"><a href="#cb71-11" aria-hidden="true" tabindex="-1"></a><span class="at">    </span><span class="fu">runtime</span><span class="kw">:</span><span class="at"> </span><span class="dv">240</span></span>
<span id="cb71-12"><a href="#cb71-12" aria-hidden="true" tabindex="-1"></a><span class="at">    </span><span class="fu">mem_mb</span><span class="kw">:</span><span class="at"> </span><span class="dv">32000</span></span>
<span id="cb71-13"><a href="#cb71-13" aria-hidden="true" tabindex="-1"></a><span class="at">    </span><span class="fu">cpus_per_task</span><span class="kw">:</span><span class="at"> </span><span class="dv">8</span></span>
<span id="cb71-14"><a href="#cb71-14" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">align_to_genome</span><span class="kw">:</span></span>
<span id="cb71-15"><a href="#cb71-15" aria-hidden="true" tabindex="-1"></a><span class="at">    </span><span class="fu">runtime</span><span class="kw">:</span><span class="at"> </span><span class="dv">120</span></span>
<span id="cb71-16"><a href="#cb71-16" aria-hidden="true" tabindex="-1"></a><span class="at">    </span><span class="fu">mem_mb</span><span class="kw">:</span><span class="at"> </span><span class="dv">24000</span></span>
<span id="cb71-17"><a href="#cb71-17" aria-hidden="true" tabindex="-1"></a><span class="at">    </span><span class="fu">cpus_per_task</span><span class="kw">:</span><span class="at"> </span><span class="dv">6</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>In this example, the <code>index_genome</code> rule will have a runtime of 240 minutes, will require 32G of RAM and 8 CPUs, while the <code>align_to_genome</code> rule will have a runtime of 120 minutes, will require 24G of RAM and 6 CPUs. Both rules will use the <code>slurm_account</code> and <code>slurm_partition</code> settings from the <code>default_resources</code> section, unless overridden in the rule-specific settings.</p>
<p>You can still define resources in the rule definition, but the values in the configuration profile will take precedence.</p>
<p>Now, when you run your Snakemake workflow with:</p>
<div class="sourceCode" id="cb72"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a><span class="ex">snakemake</span> <span class="at">--profile</span> dardel</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Snakemake will submit each job to the SLURM queue and inform you about both the local jobid and the SLURM jobid by writing something similar to this to your terminal:</p>
<pre><code>Job 0 has been submitted with SLURM jobid 37099380 (log: .snakemake/slurm_logs/rule_name/37099380.log).</code></pre>
<p>In this example the log output from the job will be in <code>.snakemake/slurm_logs/rule_name/37099380.log</code>.</p>
<p>You can read more details about running Snakemake on compute clusters in the <a href="https://snakemake.readthedocs.io/en/stable/executing/cluster.html">Snakemake docs</a>.</p>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p>2024 <a href="https://nbis.se">NBIS</a> | <a href="https://choosealicense.com/licenses/gpl-3.0/">GPL-3 License</a></p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
<p>Published with <a href="https://quarto.org/">Quarto</a> v1.4.549
</p>
</div>
  </div>
</footer>




<script src="../site_libs/quarto-html/zenscroll-min.js"></script>
</body></html>